{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "ab5fc055",
   "metadata": {},
   "source": [
    "# Prototype notebook : Querying data across all Omix Atlas\n",
    "_For internal use only_\n",
    "\n",
    "### Instructions:\n",
    "\n",
    "1. Please run all code cells one by one.\n",
    "2. User is required to enter an SQL query to search for the required datasets."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8b46c060",
   "metadata": {},
   "source": [
    "## Install polly-python"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b7908fd6",
   "metadata": {},
   "outputs": [],
   "source": [
    "!pip3 install polly-python --user"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "dd51dc51",
   "metadata": {},
   "outputs": [],
   "source": [
    "from polly.omixatlas import OmixAtlas\n",
    "import pandas as pd\n",
    "import sys\n",
    "import time\n",
    "import re\n",
    "import traceback"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0f86432f",
   "metadata": {},
   "source": [
    "## Authentication with Polly"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f59144ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "repo_client = OmixAtlas(os.environ['POLLY_REFRESH_TOKEN'])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "392db26d",
   "metadata": {},
   "source": [
    "## Get all OmixAtlas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "1611e45c",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_omix_atlas = pd.DataFrame.from_dict(repo_client.get_all_omixatlas()['data'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "0d11c282",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>repo_name</th>\n",
       "      <th>repo_id</th>\n",
       "      <th>indexes</th>\n",
       "      <th>diseases</th>\n",
       "      <th>organisms</th>\n",
       "      <th>sources</th>\n",
       "      <th>datatypes</th>\n",
       "      <th>dataset_count</th>\n",
       "      <th>disease_count</th>\n",
       "      <th>tissue_count</th>\n",
       "      <th>organism_count</th>\n",
       "      <th>cell_line_count</th>\n",
       "      <th>cell_type_count</th>\n",
       "      <th>drug_count</th>\n",
       "      <th>data_type_count</th>\n",
       "      <th>data_source_count</th>\n",
       "      <th>sample_count</th>\n",
       "      <th>normal_sample_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>gdc</td>\n",
       "      <td>1623221686703</td>\n",
       "      <td>{'gct_metadata': 'gdc_gct_metadata', 'h5ad_met...</td>\n",
       "      <td>[leukemia, myeloid, acute, precursor t-cell ly...</td>\n",
       "      <td>[homo sapiens]</td>\n",
       "      <td>[gdc]</td>\n",
       "      <td>[mirna expression, transcriptomics, copy numbe...</td>\n",
       "      <td>9869</td>\n",
       "      <td>75</td>\n",
       "      <td>41</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>5353</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>geo</td>\n",
       "      <td>9</td>\n",
       "      <td>{'gct_metadata': 'geo_gct_metadata', 'h5ad_met...</td>\n",
       "      <td>[normal, neoplasms, breast neoplasms, obesity,...</td>\n",
       "      <td>[homo sapiens, mus musculus, rattus norvegicus...</td>\n",
       "      <td>[geo]</td>\n",
       "      <td>[transcriptomics]</td>\n",
       "      <td>10000</td>\n",
       "      <td>3028</td>\n",
       "      <td>980</td>\n",
       "      <td>68</td>\n",
       "      <td>4300</td>\n",
       "      <td>780</td>\n",
       "      <td>1379</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1279490</td>\n",
       "      <td>839985</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>enterprise_atlas</td>\n",
       "      <td>1638441282192</td>\n",
       "      <td>{'gct_metadata': 'enterprise_atlas_gct_metadat...</td>\n",
       "      <td>[breast neoplasms, endometrial neoplasms, endo...</td>\n",
       "      <td>[homo sapiens, mus musculus]</td>\n",
       "      <td>[geo, proprietary]</td>\n",
       "      <td>[transcriptomics, single cell, mutation, mirna]</td>\n",
       "      <td>176</td>\n",
       "      <td>35</td>\n",
       "      <td>25</td>\n",
       "      <td>2</td>\n",
       "      <td>96</td>\n",
       "      <td>40</td>\n",
       "      <td>21</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>8952</td>\n",
       "      <td>3927</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>liveromix_atlas</td>\n",
       "      <td>1615965444377</td>\n",
       "      <td>{'gct_metadata': 'liveromix_atlas_gct_metadata...</td>\n",
       "      <td>[normal, carcinoma, hepatocellular, obesity, n...</td>\n",
       "      <td>[homo sapiens, mus musculus, rattus norvegicus...</td>\n",
       "      <td>[geo, lincs, tcga, metabolomics workbench, met...</td>\n",
       "      <td>[transcriptomics, mutation, metabolomics, sing...</td>\n",
       "      <td>6761</td>\n",
       "      <td>752</td>\n",
       "      <td>60</td>\n",
       "      <td>22</td>\n",
       "      <td>347</td>\n",
       "      <td>310</td>\n",
       "      <td>4704</td>\n",
       "      <td>12</td>\n",
       "      <td>10</td>\n",
       "      <td>143535</td>\n",
       "      <td>110202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>cbioportal</td>\n",
       "      <td>1623986995264</td>\n",
       "      <td>{'gct_metadata': 'cbioportal_gct_metadata', 'h...</td>\n",
       "      <td>[prostatic neoplasms, carcinoma, ductal, breas...</td>\n",
       "      <td>[homo sapiens]</td>\n",
       "      <td>[cbioportal]</td>\n",
       "      <td>[mutation, copy number variation, fusion, tran...</td>\n",
       "      <td>10000</td>\n",
       "      <td>330</td>\n",
       "      <td>262</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>58994</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>pcd</td>\n",
       "      <td>1622113130397</td>\n",
       "      <td>{'gct_metadata': 'pcd_gct_metadata', 'h5ad_met...</td>\n",
       "      <td>[none, adenocarcinoma, carcinoma, squamous cel...</td>\n",
       "      <td>[homo sapiens]</td>\n",
       "      <td>[pharmacodb]</td>\n",
       "      <td>[drug response]</td>\n",
       "      <td>10000</td>\n",
       "      <td>133</td>\n",
       "      <td>34</td>\n",
       "      <td>1</td>\n",
       "      <td>1399</td>\n",
       "      <td>1</td>\n",
       "      <td>749</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>653500</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>lincs</td>\n",
       "      <td>32</td>\n",
       "      <td>{'gct_metadata': 'lincs_gct_metadata', 'h5ad_m...</td>\n",
       "      <td>[prostatic neoplasms, normal, colorectal neopl...</td>\n",
       "      <td>[homo sapiens]</td>\n",
       "      <td>[geo, geo, lincs]</td>\n",
       "      <td>[transcriptomics]</td>\n",
       "      <td>10000</td>\n",
       "      <td>34</td>\n",
       "      <td>21</td>\n",
       "      <td>1</td>\n",
       "      <td>84</td>\n",
       "      <td>1</td>\n",
       "      <td>20347</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1371945</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>metabolomics</td>\n",
       "      <td>23</td>\n",
       "      <td>{'gct_metadata': 'metabolomics_gct_metadata', ...</td>\n",
       "      <td>[normal, neoplasms, diabetes mellitus, obesity...</td>\n",
       "      <td>[homo sapiens, mus musculus, rattus norvegicus...</td>\n",
       "      <td>[metabolomics workbench, metabolights, publica...</td>\n",
       "      <td>[metabolomics, lipidomics, single cell]</td>\n",
       "      <td>1777</td>\n",
       "      <td>214</td>\n",
       "      <td>134</td>\n",
       "      <td>184</td>\n",
       "      <td>24</td>\n",
       "      <td>23</td>\n",
       "      <td>67</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>90677</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>sc_data_lake</td>\n",
       "      <td>17</td>\n",
       "      <td>{'gct_metadata': 'sc_data_lake_gct_metadata', ...</td>\n",
       "      <td>[normal, neoplasms, breast neoplasms, melanoma...</td>\n",
       "      <td>[mus musculus, homo sapiens, macaca fascicular...</td>\n",
       "      <td>[geo, humancellatlas, gene expression omnibus ...</td>\n",
       "      <td>[single cell]</td>\n",
       "      <td>2915</td>\n",
       "      <td>520</td>\n",
       "      <td>414</td>\n",
       "      <td>5</td>\n",
       "      <td>69</td>\n",
       "      <td>406</td>\n",
       "      <td>80</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>207888</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          repo_name        repo_id  \\\n",
       "0               gdc  1623221686703   \n",
       "1               geo              9   \n",
       "2  enterprise_atlas  1638441282192   \n",
       "3   liveromix_atlas  1615965444377   \n",
       "4        cbioportal  1623986995264   \n",
       "5               pcd  1622113130397   \n",
       "6             lincs             32   \n",
       "7      metabolomics             23   \n",
       "8      sc_data_lake             17   \n",
       "\n",
       "                                             indexes  \\\n",
       "0  {'gct_metadata': 'gdc_gct_metadata', 'h5ad_met...   \n",
       "1  {'gct_metadata': 'geo_gct_metadata', 'h5ad_met...   \n",
       "2  {'gct_metadata': 'enterprise_atlas_gct_metadat...   \n",
       "3  {'gct_metadata': 'liveromix_atlas_gct_metadata...   \n",
       "4  {'gct_metadata': 'cbioportal_gct_metadata', 'h...   \n",
       "5  {'gct_metadata': 'pcd_gct_metadata', 'h5ad_met...   \n",
       "6  {'gct_metadata': 'lincs_gct_metadata', 'h5ad_m...   \n",
       "7  {'gct_metadata': 'metabolomics_gct_metadata', ...   \n",
       "8  {'gct_metadata': 'sc_data_lake_gct_metadata', ...   \n",
       "\n",
       "                                            diseases  \\\n",
       "0  [leukemia, myeloid, acute, precursor t-cell ly...   \n",
       "1  [normal, neoplasms, breast neoplasms, obesity,...   \n",
       "2  [breast neoplasms, endometrial neoplasms, endo...   \n",
       "3  [normal, carcinoma, hepatocellular, obesity, n...   \n",
       "4  [prostatic neoplasms, carcinoma, ductal, breas...   \n",
       "5  [none, adenocarcinoma, carcinoma, squamous cel...   \n",
       "6  [prostatic neoplasms, normal, colorectal neopl...   \n",
       "7  [normal, neoplasms, diabetes mellitus, obesity...   \n",
       "8  [normal, neoplasms, breast neoplasms, melanoma...   \n",
       "\n",
       "                                           organisms  \\\n",
       "0                                     [homo sapiens]   \n",
       "1  [homo sapiens, mus musculus, rattus norvegicus...   \n",
       "2                       [homo sapiens, mus musculus]   \n",
       "3  [homo sapiens, mus musculus, rattus norvegicus...   \n",
       "4                                     [homo sapiens]   \n",
       "5                                     [homo sapiens]   \n",
       "6                                     [homo sapiens]   \n",
       "7  [homo sapiens, mus musculus, rattus norvegicus...   \n",
       "8  [mus musculus, homo sapiens, macaca fascicular...   \n",
       "\n",
       "                                             sources  \\\n",
       "0                                              [gdc]   \n",
       "1                                              [geo]   \n",
       "2                                 [geo, proprietary]   \n",
       "3  [geo, lincs, tcga, metabolomics workbench, met...   \n",
       "4                                       [cbioportal]   \n",
       "5                                       [pharmacodb]   \n",
       "6                                  [geo, geo, lincs]   \n",
       "7  [metabolomics workbench, metabolights, publica...   \n",
       "8  [geo, humancellatlas, gene expression omnibus ...   \n",
       "\n",
       "                                           datatypes  dataset_count  \\\n",
       "0  [mirna expression, transcriptomics, copy numbe...           9869   \n",
       "1                                  [transcriptomics]          10000   \n",
       "2    [transcriptomics, single cell, mutation, mirna]            176   \n",
       "3  [transcriptomics, mutation, metabolomics, sing...           6761   \n",
       "4  [mutation, copy number variation, fusion, tran...          10000   \n",
       "5                                    [drug response]          10000   \n",
       "6                                  [transcriptomics]          10000   \n",
       "7            [metabolomics, lipidomics, single cell]           1777   \n",
       "8                                      [single cell]           2915   \n",
       "\n",
       "   disease_count  tissue_count  organism_count  cell_line_count  \\\n",
       "0             75            41               1                1   \n",
       "1           3028           980              68             4300   \n",
       "2             35            25               2               96   \n",
       "3            752            60              22              347   \n",
       "4            330           262               1                1   \n",
       "5            133            34               1             1399   \n",
       "6             34            21               1               84   \n",
       "7            214           134             184               24   \n",
       "8            520           414               5               69   \n",
       "\n",
       "   cell_type_count  drug_count  data_type_count  data_source_count  \\\n",
       "0                1           1                3                  1   \n",
       "1              780        1379                1                  1   \n",
       "2               40          21                4                  2   \n",
       "3              310        4704               12                 10   \n",
       "4                1           1                5                  1   \n",
       "5                1         749                1                  1   \n",
       "6                1       20347                1                  2   \n",
       "7               23          67                3                  4   \n",
       "8              406          80                1                  6   \n",
       "\n",
       "   sample_count  normal_sample_count  \n",
       "0          5353                    0  \n",
       "1       1279490               839985  \n",
       "2          8952                 3927  \n",
       "3        143535               110202  \n",
       "4         58994                    0  \n",
       "5        653500                    0  \n",
       "6       1371945                    0  \n",
       "7         90677                    0  \n",
       "8        207888                    0  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_omix_atlas"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "87b6069e",
   "metadata": {},
   "source": [
    "## Get all dataset level indexes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "abd3f53a",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_indexes = list(all_omix_atlas['indexes'])\n",
    "dataset_index = [ ind['files'] for ind in all_indexes ]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "03557b87",
   "metadata": {},
   "source": [
    "## Searching across all OA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "c87a881a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def query_split(query):\n",
    "    \n",
    "    \"\"\"\n",
    "    Returns N SQL queries equal to the number of input indexes\n",
    "    \n",
    "    Extended description of function.\n",
    "    \n",
    "    Parameters:\n",
    "    query (str): input SQL query to search across any given omixatlases\n",
    "    \n",
    "    Returns:\n",
    "    q_list (list): list of SQL queries where each query queries on a single index only.\n",
    "    \n",
    "    \"\"\"\n",
    "    \n",
    "    index_search = re.search(\"from(.*)where\", query, re.IGNORECASE)\n",
    "    index_list = index_search.group(1).split(',')\n",
    "    index_list = [index.strip() for index in index_list]\n",
    "    \n",
    "    const_string1 = query.split(index_list[0]) #Extract constant SQL query\n",
    "    const_string2 = query.split(index_list[-1]) #Extract constant SQL query\n",
    "    \n",
    "    q_list = []\n",
    "    \n",
    "    for index in index_list:\n",
    "        q = f\"{const_string1[0]}{index}{const_string2[-1]}\"\n",
    "        q_list.append(q)\n",
    "    \n",
    "    return q_list\n",
    "    \n",
    "\n",
    "def for_loop(query_list):\n",
    "    \n",
    "    \"\"\"\n",
    "    Returns a table containing dataset level metadata for all datasets based on the input SQL query\n",
    "    \n",
    "    Extended description of function.\n",
    "    \n",
    "    Parameters:\n",
    "    query_list (list): list of input SQL queries where each query is used to search across a single omixatlas\n",
    "    \n",
    "    Returns:\n",
    "    df_result (dataframe): DataFrame containing dataset level metadata of the resulting datasets.\n",
    "    \n",
    "    \"\"\"\n",
    "\n",
    "    df_list = []\n",
    "    \n",
    "    for query in query_list:\n",
    "        all_df = repo_client.query_metadata(query)\n",
    "        if all_df is not None:\n",
    "            df_list.append(all_df)\n",
    "    \n",
    "    df_result = pd.concat(df_list) #Merge all results into a single table\n",
    "    \n",
    "    return df_result\n",
    "\n",
    "def multi_index(query):\n",
    "\n",
    "    all_df = repo_client.query_metadata(query)\n",
    "    return all_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d78daf62",
   "metadata": {},
   "source": [
    "## Input SQL query\n",
    "\n",
    "#### List of available dataset indexes\n",
    "\n",
    "1. gdc_files\n",
    "2. geo_files\n",
    "3. enterprise_atlas_files\n",
    "4. liveromix_atlas_files\n",
    "5. cbioportal_files\n",
    "6. pcd_files\n",
    "7. lincs_files\n",
    "8. metabolomics_files\n",
    "9. sc_data_lake_files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "aa7cefae",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Caught this error: VerificationException()\n",
      "{'query_used': \"SELECT * from pcd_files WHERE disease = 'Carcinoma, Hepatocellular' LIMIT 2000\", 'message': 'Showing 1 - 603 of 603 matching results'}\n",
      "{'query_used': \"SELECT * from geo_files WHERE disease = 'Carcinoma, Hepatocellular' LIMIT 2000\", 'message': 'Showing 1 - 942 of 942 matching results'}\n",
      "--- 45.4629647731781 seconds ---\n"
     ]
    }
   ],
   "source": [
    "if __name__ == \"__main__\":\n",
    "    \n",
    "    sql = \"SELECT * from pcd_files, geo_files WHERE disease = 'Carcinoma, Hepatocellular' LIMIT 2000\"\n",
    "    \n",
    "    start_time = time.time()\n",
    "    try:\n",
    "        result_df = multi_index(sql)\n",
    "    except Exception as e:\n",
    "        print('Caught this Exception: ' + repr(e))\n",
    "        query_list = query_split(sql)\n",
    "        result_df = for_loop(query_list)\n",
    "    \n",
    "    print(\"--- %s seconds ---\" % (time.time() - start_time))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "e80b1f1b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dataset_source</th>\n",
       "      <th>dataset_id</th>\n",
       "      <th>description</th>\n",
       "      <th>kw_cell_line</th>\n",
       "      <th>organism</th>\n",
       "      <th>year</th>\n",
       "      <th>operation</th>\n",
       "      <th>kw_source</th>\n",
       "      <th>is_public</th>\n",
       "      <th>data_repository</th>\n",
       "      <th>...</th>\n",
       "      <th>kw_strain</th>\n",
       "      <th>file_type</th>\n",
       "      <th>source_process</th>\n",
       "      <th>processing</th>\n",
       "      <th>drug</th>\n",
       "      <th>author</th>\n",
       "      <th>abstract</th>\n",
       "      <th>overall_design</th>\n",
       "      <th>summary</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PharmacoDB</td>\n",
       "      <td>17-AAG_haematopoietic_and_lymphoid_tissue_GDSC...</td>\n",
       "      <td>dose response data for 17-AAG drug and cell li...</td>\n",
       "      <td>[BC-1, Jurkat, GR-ST, AMO1, Ku812, K-562, WSU-...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>2018</td>\n",
       "      <td>{'is_normalized': 'false', 'normalized_type': ...</td>\n",
       "      <td>GDSC1000</td>\n",
       "      <td>true</td>\n",
       "      <td>PharmacoDB</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PharmacoDB</td>\n",
       "      <td>16-beta-bromoandrosterone_haematopoietic_and_l...</td>\n",
       "      <td>dose response data for 16-beta-bromoandrostero...</td>\n",
       "      <td>[Jurkat, MHH-CALL-4, JK-1, HPB-ALL, Ku812, KMS...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>2018</td>\n",
       "      <td>{'is_normalized': 'false', 'normalized_type': ...</td>\n",
       "      <td>CTRPv2</td>\n",
       "      <td>true</td>\n",
       "      <td>PharmacoDB</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PharmacoDB</td>\n",
       "      <td>5-FU_haematopoietic_and_lymphoid_tissue_CTRPv2...</td>\n",
       "      <td>dose response data for 5-FU drug and cell line...</td>\n",
       "      <td>[Jurkat, MHH-CALL-4, JK-1, HPB-ALL, Ku812, KMS...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>2018</td>\n",
       "      <td>{'is_normalized': 'false', 'normalized_type': ...</td>\n",
       "      <td>CTRPv2</td>\n",
       "      <td>true</td>\n",
       "      <td>PharmacoDB</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PharmacoDB</td>\n",
       "      <td>ABT-199_haematopoietic_and_lymphoid_tissue_CTR...</td>\n",
       "      <td>dose response data for ABT-199 drug and cell l...</td>\n",
       "      <td>[MHH-CALL-4, JK-1, HPB-ALL, AMO1, ME1, RPMI-82...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>2018</td>\n",
       "      <td>{'is_normalized': 'false', 'normalized_type': ...</td>\n",
       "      <td>CTRPv2</td>\n",
       "      <td>true</td>\n",
       "      <td>PharmacoDB</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PharmacoDB</td>\n",
       "      <td>AP-24534_haematopoietic_and_lymphoid_tissue_GD...</td>\n",
       "      <td>dose response data for AP-24534 drug and cell ...</td>\n",
       "      <td>[SCC-3, BC-1, GR-ST, AMO1, Ku812, K-562, WSU-N...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>2018</td>\n",
       "      <td>{'is_normalized': 'false', 'normalized_type': ...</td>\n",
       "      <td>GDSC1000</td>\n",
       "      <td>true</td>\n",
       "      <td>PharmacoDB</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>937</th>\n",
       "      <td>GEO</td>\n",
       "      <td>GSE82177_GPL11154</td>\n",
       "      <td>Human liver RNA-seq data corresponding to unin...</td>\n",
       "      <td>[None]</td>\n",
       "      <td>[Homo sapiens]</td>\n",
       "      <td>2018</td>\n",
       "      <td>{'is_normalized': 'true', 'batch_corrected_var...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>true</td>\n",
       "      <td>geo</td>\n",
       "      <td>...</td>\n",
       "      <td>[None]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>938</th>\n",
       "      <td>GEO</td>\n",
       "      <td>GSE84186_GPL11154</td>\n",
       "      <td>Single-nucleotide-resolution mapping of HBV pr...</td>\n",
       "      <td>[Hep-G2/2.2.15, HepAD38]</td>\n",
       "      <td>[Homo sapiens]</td>\n",
       "      <td>2018</td>\n",
       "      <td>{'is_normalized': 'true', 'batch_corrected_var...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>true</td>\n",
       "      <td>geo</td>\n",
       "      <td>...</td>\n",
       "      <td>[None]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>939</th>\n",
       "      <td>GEO</td>\n",
       "      <td>GSE87240_GPL11154</td>\n",
       "      <td>S-adenosyl-methionine treatment selectively bl...</td>\n",
       "      <td>[None]</td>\n",
       "      <td>[Homo sapiens]</td>\n",
       "      <td>2018</td>\n",
       "      <td>{'is_normalized': 'true', 'batch_corrected_var...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>true</td>\n",
       "      <td>geo</td>\n",
       "      <td>...</td>\n",
       "      <td>[None]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>940</th>\n",
       "      <td>GEO</td>\n",
       "      <td>GSE85427_GPL13112</td>\n",
       "      <td>Insulin-like Growth Factor Binding Protein-7 (...</td>\n",
       "      <td>[None]</td>\n",
       "      <td>[Mus musculus]</td>\n",
       "      <td>Oct 01 2019</td>\n",
       "      <td>{'is_normalized': 'true', 'batch_corrected_var...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>true</td>\n",
       "      <td>geo</td>\n",
       "      <td>...</td>\n",
       "      <td>[C57BL/6]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Yidong,,Chen</td>\n",
       "      <td>Insulin like growth factor binding protein-7 (...</td>\n",
       "      <td>Three replicates of RNAs derived from WT and I...</td>\n",
       "      <td>Insulin like growth factor binding protein-7 (...</td>\n",
       "      <td>Expression profiling by high throughput sequen...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>941</th>\n",
       "      <td>GEO</td>\n",
       "      <td>GSE97412_GPL18573</td>\n",
       "      <td>Expression analysis of INO80 complex in HepG2 ...</td>\n",
       "      <td>[Hep-G2]</td>\n",
       "      <td>[Homo sapiens]</td>\n",
       "      <td>Apr 01 2020</td>\n",
       "      <td>{'is_normalized': 'true', 'batch_corrected_var...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>true</td>\n",
       "      <td>geo</td>\n",
       "      <td>...</td>\n",
       "      <td>[None]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Jesse,,Raab</td>\n",
       "      <td>We carried out two sets of experiments to unde...</td>\n",
       "      <td>HepG2 cells were treated with siRNA against IN...</td>\n",
       "      <td>We carried out two sets of experiments to unde...</td>\n",
       "      <td>Expression profiling by high throughput sequen...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1545 rows Ã— 42 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    dataset_source                                         dataset_id  \\\n",
       "0       PharmacoDB  17-AAG_haematopoietic_and_lymphoid_tissue_GDSC...   \n",
       "1       PharmacoDB  16-beta-bromoandrosterone_haematopoietic_and_l...   \n",
       "2       PharmacoDB  5-FU_haematopoietic_and_lymphoid_tissue_CTRPv2...   \n",
       "3       PharmacoDB  ABT-199_haematopoietic_and_lymphoid_tissue_CTR...   \n",
       "4       PharmacoDB  AP-24534_haematopoietic_and_lymphoid_tissue_GD...   \n",
       "..             ...                                                ...   \n",
       "937            GEO                                  GSE82177_GPL11154   \n",
       "938            GEO                                  GSE84186_GPL11154   \n",
       "939            GEO                                  GSE87240_GPL11154   \n",
       "940            GEO                                  GSE85427_GPL13112   \n",
       "941            GEO                                  GSE97412_GPL18573   \n",
       "\n",
       "                                           description  \\\n",
       "0    dose response data for 17-AAG drug and cell li...   \n",
       "1    dose response data for 16-beta-bromoandrostero...   \n",
       "2    dose response data for 5-FU drug and cell line...   \n",
       "3    dose response data for ABT-199 drug and cell l...   \n",
       "4    dose response data for AP-24534 drug and cell ...   \n",
       "..                                                 ...   \n",
       "937  Human liver RNA-seq data corresponding to unin...   \n",
       "938  Single-nucleotide-resolution mapping of HBV pr...   \n",
       "939  S-adenosyl-methionine treatment selectively bl...   \n",
       "940  Insulin-like Growth Factor Binding Protein-7 (...   \n",
       "941  Expression analysis of INO80 complex in HepG2 ...   \n",
       "\n",
       "                                          kw_cell_line        organism  \\\n",
       "0    [BC-1, Jurkat, GR-ST, AMO1, Ku812, K-562, WSU-...    Homo sapiens   \n",
       "1    [Jurkat, MHH-CALL-4, JK-1, HPB-ALL, Ku812, KMS...    Homo sapiens   \n",
       "2    [Jurkat, MHH-CALL-4, JK-1, HPB-ALL, Ku812, KMS...    Homo sapiens   \n",
       "3    [MHH-CALL-4, JK-1, HPB-ALL, AMO1, ME1, RPMI-82...    Homo sapiens   \n",
       "4    [SCC-3, BC-1, GR-ST, AMO1, Ku812, K-562, WSU-N...    Homo sapiens   \n",
       "..                                                 ...             ...   \n",
       "937                                             [None]  [Homo sapiens]   \n",
       "938                           [Hep-G2/2.2.15, HepAD38]  [Homo sapiens]   \n",
       "939                                             [None]  [Homo sapiens]   \n",
       "940                                             [None]  [Mus musculus]   \n",
       "941                                           [Hep-G2]  [Homo sapiens]   \n",
       "\n",
       "            year                                          operation kw_source  \\\n",
       "0           2018  {'is_normalized': 'false', 'normalized_type': ...  GDSC1000   \n",
       "1           2018  {'is_normalized': 'false', 'normalized_type': ...    CTRPv2   \n",
       "2           2018  {'is_normalized': 'false', 'normalized_type': ...    CTRPv2   \n",
       "3           2018  {'is_normalized': 'false', 'normalized_type': ...    CTRPv2   \n",
       "4           2018  {'is_normalized': 'false', 'normalized_type': ...  GDSC1000   \n",
       "..           ...                                                ...       ...   \n",
       "937         2018  {'is_normalized': 'true', 'batch_corrected_var...       NaN   \n",
       "938         2018  {'is_normalized': 'true', 'batch_corrected_var...       NaN   \n",
       "939         2018  {'is_normalized': 'true', 'batch_corrected_var...       NaN   \n",
       "940  Oct 01 2019  {'is_normalized': 'true', 'batch_corrected_var...       NaN   \n",
       "941  Apr 01 2020  {'is_normalized': 'true', 'batch_corrected_var...       NaN   \n",
       "\n",
       "    is_public data_repository  ...  kw_strain file_type source_process  \\\n",
       "0        true      PharmacoDB  ...        NaN       NaN            NaN   \n",
       "1        true      PharmacoDB  ...        NaN       NaN            NaN   \n",
       "2        true      PharmacoDB  ...        NaN       NaN            NaN   \n",
       "3        true      PharmacoDB  ...        NaN       NaN            NaN   \n",
       "4        true      PharmacoDB  ...        NaN       NaN            NaN   \n",
       "..        ...             ...  ...        ...       ...            ...   \n",
       "937      true             geo  ...     [None]       NaN            NaN   \n",
       "938      true             geo  ...     [None]       NaN            NaN   \n",
       "939      true             geo  ...     [None]       NaN            NaN   \n",
       "940      true             geo  ...  [C57BL/6]       NaN            NaN   \n",
       "941      true             geo  ...     [None]       NaN            NaN   \n",
       "\n",
       "    processing drug        author  \\\n",
       "0          NaN  NaN           NaN   \n",
       "1          NaN  NaN           NaN   \n",
       "2          NaN  NaN           NaN   \n",
       "3          NaN  NaN           NaN   \n",
       "4          NaN  NaN           NaN   \n",
       "..         ...  ...           ...   \n",
       "937        NaN  NaN           NaN   \n",
       "938        NaN  NaN           NaN   \n",
       "939        NaN  NaN           NaN   \n",
       "940        NaN  NaN  Yidong,,Chen   \n",
       "941        NaN  NaN   Jesse,,Raab   \n",
       "\n",
       "                                              abstract  \\\n",
       "0                                                  NaN   \n",
       "1                                                  NaN   \n",
       "2                                                  NaN   \n",
       "3                                                  NaN   \n",
       "4                                                  NaN   \n",
       "..                                                 ...   \n",
       "937                                                NaN   \n",
       "938                                                NaN   \n",
       "939                                                NaN   \n",
       "940  Insulin like growth factor binding protein-7 (...   \n",
       "941  We carried out two sets of experiments to unde...   \n",
       "\n",
       "                                        overall_design  \\\n",
       "0                                                  NaN   \n",
       "1                                                  NaN   \n",
       "2                                                  NaN   \n",
       "3                                                  NaN   \n",
       "4                                                  NaN   \n",
       "..                                                 ...   \n",
       "937                                                NaN   \n",
       "938                                                NaN   \n",
       "939                                                NaN   \n",
       "940  Three replicates of RNAs derived from WT and I...   \n",
       "941  HepG2 cells were treated with siRNA against IN...   \n",
       "\n",
       "                                               summary  \\\n",
       "0                                                  NaN   \n",
       "1                                                  NaN   \n",
       "2                                                  NaN   \n",
       "3                                                  NaN   \n",
       "4                                                  NaN   \n",
       "..                                                 ...   \n",
       "937                                                NaN   \n",
       "938                                                NaN   \n",
       "939                                                NaN   \n",
       "940  Insulin like growth factor binding protein-7 (...   \n",
       "941  We carried out two sets of experiments to unde...   \n",
       "\n",
       "                                                  type  \n",
       "0                                                  NaN  \n",
       "1                                                  NaN  \n",
       "2                                                  NaN  \n",
       "3                                                  NaN  \n",
       "4                                                  NaN  \n",
       "..                                                 ...  \n",
       "937                                                NaN  \n",
       "938                                                NaN  \n",
       "939                                                NaN  \n",
       "940  Expression profiling by high throughput sequen...  \n",
       "941  Expression profiling by high throughput sequen...  \n",
       "\n",
       "[1545 rows x 42 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c7e0c272",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "general",
   "language": "python",
   "name": "general"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
