{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "colab_type": "text",
        "id": "view-in-github"
      },
      "source": [
        "<a href=\"https://polly.elucidata.io/manage/workspaces?action=open_polly_notebook&source=github&path=path_place_holder&kernel=elucidata/Python 3.10&machine=medium\" target=\"_parent\"><img src=\"https://elucidatainc.github.io/PublicAssets/open_polly.svg\" alt=\"Open in Polly\"/></a>\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "# Install Polly Python"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 1,
      "metadata": {},
      "outputs": [],
      "source": [
        "# please do not modify\n",
        "from IPython.display import display_html\n",
        "def restartkernel() :\n",
        "    display_html(\"<script>Jupyter.notebook.kernel.restart()</script>\",raw=True)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 2,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "/bin/bash: warning: setlocale: LC_ALL: cannot change locale (en_US.UTF-8)\n",
            "Collecting polly-python\n",
            "  Downloading polly_python-0.2.2-py3-none-any.whl (69 kB)\n",
            "\u001b[K     |████████████████████████████████| 69 kB 4.9 MB/s eta 0:00:011\n",
            "\u001b[?25hCollecting pytz==2021.1\n",
            "  Downloading pytz-2021.1-py2.py3-none-any.whl (510 kB)\n",
            "\u001b[K     |████████████████████████████████| 510 kB 5.6 MB/s eta 0:00:011\n",
            "\u001b[?25hCollecting joblib\n",
            "  Downloading joblib-1.2.0-py3-none-any.whl (297 kB)\n",
            "\u001b[K     |████████████████████████████████| 297 kB 136.1 MB/s eta 0:00:01\n",
            "\u001b[?25hCollecting cmapPy\n",
            "  Downloading cmapPy-4.0.1-py2.py3-none-any.whl (150 kB)\n",
            "\u001b[K     |████████████████████████████████| 150 kB 131.8 MB/s eta 0:00:01\n",
            "\u001b[?25hRequirement already satisfied: idna==2.10 in /usr/local/lib/python3.10/site-packages (from polly-python) (2.10)\n",
            "Requirement already satisfied: python-dateutil in /usr/local/lib/python3.10/site-packages (from polly-python) (2.8.1)\n",
            "Collecting certifi==2021.10.8\n",
            "  Downloading certifi-2021.10.8-py2.py3-none-any.whl (149 kB)\n",
            "\u001b[K     |████████████████████████████████| 149 kB 136.8 MB/s eta 0:00:01\n",
            "\u001b[?25hCollecting datapane\n",
            "  Downloading datapane-0.15.3-py3-none-any.whl (3.5 MB)\n",
            "\u001b[K     |████████████████████████████████| 3.5 MB 139.9 MB/s eta 0:00:01\n",
            "\u001b[?25hCollecting rst2txt\n",
            "  Downloading rst2txt-1.1.0-py2.py3-none-any.whl (12 kB)\n",
            "Collecting python-magic==0.4.24\n",
            "  Downloading python_magic-0.4.24-py2.py3-none-any.whl (12 kB)\n",
            "Collecting plotly\n",
            "  Downloading plotly-5.10.0-py2.py3-none-any.whl (15.2 MB)\n",
            "\u001b[K     |████████████████████████████████| 15.2 MB 127.9 MB/s eta 0:00:01\n",
            "\u001b[?25hCollecting beautifulsoup4\n",
            "  Downloading beautifulsoup4-4.11.1-py3-none-any.whl (128 kB)\n",
            "\u001b[K     |████████████████████████████████| 128 kB 149.6 MB/s eta 0:00:01\n",
            "\u001b[?25hCollecting elucidatacloudpathlib==0.6.6\n",
            "  Downloading elucidatacloudpathlib-0.6.6-py3-none-any.whl (50 kB)\n",
            "\u001b[K     |████████████████████████████████| 50 kB 18.6 MB/s  eta 0:00:01\n",
            "\u001b[?25hCollecting Deprecated\n",
            "  Downloading Deprecated-1.2.13-py2.py3-none-any.whl (9.6 kB)\n",
            "Requirement already satisfied: urllib3==1.26.6 in /usr/local/lib/python3.10/site-packages (from polly-python) (1.26.6)\n",
            "Requirement already satisfied: six==1.16.0 in /usr/local/lib/python3.10/site-packages (from polly-python) (1.16.0)\n",
            "Requirement already satisfied: tqdm in /usr/local/lib/python3.10/site-packages (from polly-python) (4.64.0)\n",
            "Requirement already satisfied: pandas in /usr/local/lib/python3.10/site-packages (from polly-python) (1.4.1)\n",
            "Collecting Cerberus==1.3.2\n",
            "  Downloading Cerberus-1.3.2.tar.gz (52 kB)\n",
            "\u001b[K     |████████████████████████████████| 52 kB 3.7 MB/s s eta 0:00:01\n",
            "\u001b[?25hRequirement already satisfied: botocore>=1.20.73 in /usr/local/lib/python3.10/site-packages (from polly-python) (1.24.19)\n",
            "Requirement already satisfied: boto3>=1.17.73 in /usr/local/lib/python3.10/site-packages (from polly-python) (1.21.19)\n",
            "Collecting requests==2.25.1\n",
            "  Downloading requests-2.25.1-py2.py3-none-any.whl (61 kB)\n",
            "\u001b[K     |████████████████████████████████| 61 kB 21.1 MB/s  eta 0:00:01\n",
            "\u001b[?25hRequirement already satisfied: chardet==4.0.0 in /usr/local/lib/python3.10/site-packages (from polly-python) (4.0.0)\n",
            "Collecting pydantic==1.9.1\n",
            "  Downloading pydantic-1.9.1-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (12.0 MB)\n",
            "\u001b[K     |████████████████████████████████| 12.0 MB 112.2 MB/s eta 0:00:01\n",
            "\u001b[?25hCollecting sqlparse\n",
            "  Downloading sqlparse-0.4.3-py3-none-any.whl (42 kB)\n",
            "\u001b[K     |████████████████████████████████| 42 kB 3.8 MB/s s eta 0:00:01\n",
            "\u001b[?25hCollecting postpy2==0.0.6\n",
            "  Downloading postpy2-0.0.6-py3-none-any.whl (17 kB)\n",
            "Collecting pytest\n",
            "  Downloading pytest-7.1.3-py3-none-any.whl (298 kB)\n",
            "\u001b[K     |████████████████████████████████| 298 kB 146.7 MB/s eta 0:00:01\n",
            "\u001b[?25hCollecting retrying==1.3.3\n",
            "  Downloading retrying-1.3.3.tar.gz (10 kB)\n",
            "Requirement already satisfied: setuptools in /usr/local/lib/python3.10/site-packages (from Cerberus==1.3.2->polly-python) (58.1.0)\n",
            "Collecting typing-extensions>=3.7.4.3\n",
            "  Downloading typing_extensions-4.4.0-py3-none-any.whl (26 kB)\n",
            "Requirement already satisfied: s3transfer<0.6.0,>=0.5.0 in /usr/local/lib/python3.10/site-packages (from boto3>=1.17.73->polly-python) (0.5.2)\n",
            "Requirement already satisfied: jmespath<1.0.0,>=0.7.1 in /usr/local/lib/python3.10/site-packages (from boto3>=1.17.73->polly-python) (0.10.0)\n",
            "Collecting soupsieve>1.2\n",
            "  Downloading soupsieve-2.3.2.post1-py3-none-any.whl (37 kB)\n",
            "Collecting h5py>=2.6.0\n",
            "  Downloading h5py-3.7.0-cp310-cp310-manylinux_2_12_x86_64.manylinux2010_x86_64.whl (4.5 MB)\n",
            "\u001b[K     |████████████████████████████████| 4.5 MB 117.5 MB/s eta 0:00:01\n",
            "\u001b[?25hRequirement already satisfied: numpy>=1.11.2 in /usr/local/lib/python3.10/site-packages (from cmapPy->polly-python) (1.22.3)\n",
            "Requirement already satisfied: jsonschema<5.0.0,>=3.2.0 in /usr/local/lib/python3.10/site-packages (from datapane->polly-python) (4.2.1)\n",
            "Collecting datacommons-pandas<0.0.4,>=0.0.3\n",
            "  Downloading datacommons_pandas-0.0.3-py3-none-any.whl (45 kB)\n",
            "\u001b[K     |████████████████████████████████| 45 kB 8.4 MB/s s eta 0:00:01\n",
            "\u001b[?25hCollecting munch<3.0.0,>=2.3.0\n",
            "  Downloading munch-2.5.0-py2.py3-none-any.whl (10 kB)\n",
            "Requirement already satisfied: packaging<22.0.0,>=20.0.0 in /usr/local/lib/python3.10/site-packages (from datapane->polly-python) (21.3)\n",
            "Collecting dulwich<0.21.0,>=0.20.0\n",
            "  Downloading dulwich-0.20.46-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (497 kB)\n",
            "\u001b[K     |████████████████████████████████| 497 kB 143.2 MB/s eta 0:00:01\n",
            "\u001b[?25hCollecting importlib_resources<6.0.0,>=3.0.0\n",
            "  Downloading importlib_resources-5.10.0-py3-none-any.whl (34 kB)\n",
            "Collecting lxml<5.0.0,>=4.0.0\n",
            "  Downloading lxml-4.9.1-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.manylinux_2_24_x86_64.whl (6.9 MB)\n",
            "\u001b[K     |████████████████████████████████| 6.9 MB 102.7 MB/s eta 0:00:01\n",
            "\u001b[?25hCollecting datacommons<2.0.0,>=1.4.3\n",
            "  Downloading datacommons-1.4.3-py3-none-any.whl (46 kB)\n",
            "\u001b[K     |████████████████████████████████| 46 kB 9.8 MB/s s eta 0:00:01\n",
            "\u001b[?25hCollecting dacite<2.0.0,>=1.0.2\n",
            "  Downloading dacite-1.6.0-py3-none-any.whl (12 kB)\n",
            "Collecting validators<0.21.0,>=0.18.0\n",
            "  Downloading validators-0.20.0.tar.gz (30 kB)\n",
            "Collecting stringcase<2.0.0,>=1.2.0\n",
            "  Downloading stringcase-1.2.0.tar.gz (3.0 kB)\n",
            "Requirement already satisfied: Jinja2<4.0.0,>=3.0.0 in /usr/local/lib/python3.10/site-packages (from datapane->polly-python) (3.0.3)\n",
            "Collecting micawber>=0.5.0\n",
            "  Downloading micawber-0.5.4.tar.gz (18 kB)\n",
            "Collecting vega-datasets<1.0.0,>=0.9.0\n",
            "  Downloading vega_datasets-0.9.0-py3-none-any.whl (210 kB)\n",
            "\u001b[K     |████████████████████████████████| 210 kB 141.2 MB/s eta 0:00:01\n",
            "\u001b[?25hCollecting boltons<22.0.0,>=20.0.0\n",
            "  Downloading boltons-21.0.0-py2.py3-none-any.whl (193 kB)\n",
            "\u001b[K     |████████████████████████████████| 193 kB 129.1 MB/s eta 0:00:01\n",
            "\u001b[?25hCollecting glom<23.0.0,>=20.11.0\n",
            "  Downloading glom-22.1.0-py2.py3-none-any.whl (100 kB)\n",
            "\u001b[K     |████████████████████████████████| 100 kB 28.8 MB/s eta 0:00:01\n",
            "\u001b[?25hCollecting requests-toolbelt<0.10.0,>=0.9.1\n",
            "  Downloading requests_toolbelt-0.9.1-py2.py3-none-any.whl (54 kB)\n",
            "\u001b[K     |████████████████████████████████| 54 kB 7.4 MB/s s eta 0:00:01\n",
            "\u001b[?25hRequirement already satisfied: tabulate<0.9.0,>=0.8.0 in /usr/local/lib/python3.10/site-packages (from datapane->polly-python) (0.8.9)\n",
            "Requirement already satisfied: nbconvert<7.0.0,>=6.1.0 in /usr/local/lib/python3.10/site-packages (from datapane->polly-python) (6.3.0)\n",
            "Collecting colorlog<7.0.0,>=4.1.0\n",
            "  Downloading colorlog-6.7.0-py2.py3-none-any.whl (11 kB)\n",
            "Collecting click<9.0.0,>=7.1.0\n",
            "  Downloading click-8.1.3-py3-none-any.whl (96 kB)\n",
            "\u001b[K     |████████████████████████████████| 96 kB 16.9 MB/s  eta 0:00:01\n",
            "\u001b[?25hCollecting click-spinner<0.2.0,>=0.1.8\n",
            "  Downloading click_spinner-0.1.10-py2.py3-none-any.whl (4.0 kB)\n",
            "Collecting dominate<3.0.0,>=2.4.0\n",
            "  Downloading dominate-2.7.0-py2.py3-none-any.whl (29 kB)\n",
            "Collecting altair<5.0.0,>=4.0.0\n",
            "  Downloading altair-4.2.0-py3-none-any.whl (812 kB)\n",
            "\u001b[K     |████████████████████████████████| 812 kB 135.4 MB/s eta 0:00:01\n",
            "\u001b[?25hCollecting posthog<3.0.0,>=1.4.0\n",
            "  Downloading posthog-2.1.2-py2.py3-none-any.whl (32 kB)\n",
            "Requirement already satisfied: PyYAML<7.0.0,>=5.4.0 in /usr/local/lib/python3.10/site-packages (from datapane->polly-python) (5.4.1)\n"
          ]
        },
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Collecting pyarrow<7.0.0,>=3.0.0\n",
            "  Downloading pyarrow-6.0.1-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (25.6 MB)\n",
            "\u001b[K     |████████████████████████████████| 25.6 MB 129.0 MB/s eta 0:00:01\u001b[K     |████████████████████████████████| 25.6 MB 129.0 MB/s \n",
            "\u001b[?25hCollecting toolz<0.13.0,>=0.11.0\n",
            "  Downloading toolz-0.12.0-py3-none-any.whl (55 kB)\n",
            "\u001b[K     |████████████████████████████████| 55 kB 12.4 MB/s  eta 0:00:01\n",
            "\u001b[?25hCollecting furl<3.0.0,>=2.0.0\n",
            "  Downloading furl-2.1.3-py2.py3-none-any.whl (20 kB)\n",
            "Requirement already satisfied: entrypoints in /usr/local/lib/python3.10/site-packages (from altair<5.0.0,>=4.0.0->datapane->polly-python) (0.3)\n",
            "Collecting orderedmultidict>=1.0.1\n",
            "  Downloading orderedmultidict-1.0.1-py2.py3-none-any.whl (11 kB)\n",
            "Requirement already satisfied: attrs in /usr/local/lib/python3.10/site-packages (from glom<23.0.0,>=20.11.0->datapane->polly-python) (21.2.0)\n",
            "Collecting face>=20.1.0\n",
            "  Downloading face-22.0.0-py3-none-any.whl (54 kB)\n",
            "\u001b[K     |████████████████████████████████| 54 kB 9.3 MB/s s eta 0:00:01\n",
            "\u001b[?25hRequirement already satisfied: MarkupSafe>=2.0 in /usr/local/lib/python3.10/site-packages (from Jinja2<4.0.0,>=3.0.0->datapane->polly-python) (2.0.1)\n",
            "Requirement already satisfied: pyrsistent!=0.17.0,!=0.17.1,!=0.17.2,>=0.14.0 in /usr/local/lib/python3.10/site-packages (from jsonschema<5.0.0,>=3.2.0->datapane->polly-python) (0.18.0)\n",
            "Requirement already satisfied: nbformat>=4.4 in /usr/local/lib/python3.10/site-packages (from nbconvert<7.0.0,>=6.1.0->datapane->polly-python) (5.1.3)\n",
            "Requirement already satisfied: traitlets>=5.0 in /usr/local/lib/python3.10/site-packages (from nbconvert<7.0.0,>=6.1.0->datapane->polly-python) (5.1.1)\n",
            "Requirement already satisfied: mistune<2,>=0.8.1 in /usr/local/lib/python3.10/site-packages (from nbconvert<7.0.0,>=6.1.0->datapane->polly-python) (0.8.4)\n",
            "Requirement already satisfied: bleach in /usr/local/lib/python3.10/site-packages (from nbconvert<7.0.0,>=6.1.0->datapane->polly-python) (4.1.0)\n",
            "Requirement already satisfied: testpath in /usr/local/lib/python3.10/site-packages (from nbconvert<7.0.0,>=6.1.0->datapane->polly-python) (0.5.0)\n",
            "Requirement already satisfied: jupyterlab-pygments in /usr/local/lib/python3.10/site-packages (from nbconvert<7.0.0,>=6.1.0->datapane->polly-python) (0.2.2)\n",
            "Requirement already satisfied: jupyter-core in /usr/local/lib/python3.10/site-packages (from nbconvert<7.0.0,>=6.1.0->datapane->polly-python) (4.9.1)\n",
            "Requirement already satisfied: pandocfilters>=1.4.1 in /usr/local/lib/python3.10/site-packages (from nbconvert<7.0.0,>=6.1.0->datapane->polly-python) (1.5.0)\n",
            "Requirement already satisfied: defusedxml in /usr/local/lib/python3.10/site-packages (from nbconvert<7.0.0,>=6.1.0->datapane->polly-python) (0.7.1)\n",
            "Requirement already satisfied: nbclient<0.6.0,>=0.5.0 in /usr/local/lib/python3.10/site-packages (from nbconvert<7.0.0,>=6.1.0->datapane->polly-python) (0.5.13)\n",
            "Requirement already satisfied: pygments>=2.4.1 in /usr/local/lib/python3.10/site-packages (from nbconvert<7.0.0,>=6.1.0->datapane->polly-python) (2.10.0)\n",
            "Requirement already satisfied: nest-asyncio in /usr/local/lib/python3.10/site-packages (from nbclient<0.6.0,>=0.5.0->nbconvert<7.0.0,>=6.1.0->datapane->polly-python) (1.5.5)\n",
            "Requirement already satisfied: jupyter-client>=6.1.5 in /usr/local/lib/python3.10/site-packages (from nbclient<0.6.0,>=0.5.0->nbconvert<7.0.0,>=6.1.0->datapane->polly-python) (6.1.12)\n",
            "Requirement already satisfied: pyzmq>=13 in /usr/local/lib/python3.10/site-packages (from jupyter-client>=6.1.5->nbclient<0.6.0,>=0.5.0->nbconvert<7.0.0,>=6.1.0->datapane->polly-python) (22.3.0)\n",
            "Requirement already satisfied: tornado>=4.1 in /usr/local/lib/python3.10/site-packages (from jupyter-client>=6.1.5->nbclient<0.6.0,>=0.5.0->nbconvert<7.0.0,>=6.1.0->datapane->polly-python) (6.1)\n",
            "Requirement already satisfied: ipython-genutils in /usr/local/lib/python3.10/site-packages (from nbformat>=4.4->nbconvert<7.0.0,>=6.1.0->datapane->polly-python) (0.2.0)\n",
            "Requirement already satisfied: pyparsing!=3.0.5,>=2.0.2 in /usr/local/lib/python3.10/site-packages (from packaging<22.0.0,>=20.0.0->datapane->polly-python) (3.0.6)\n",
            "Collecting backoff<2.0.0,>=1.10.0\n",
            "  Downloading backoff-1.11.1-py2.py3-none-any.whl (13 kB)\n",
            "Collecting monotonic>=1.5\n",
            "  Downloading monotonic-1.6-py2.py3-none-any.whl (8.2 kB)\n",
            "Requirement already satisfied: decorator>=3.4.0 in /usr/local/lib/python3.10/site-packages (from validators<0.21.0,>=0.18.0->datapane->polly-python) (5.1.0)\n",
            "Requirement already satisfied: webencodings in /usr/local/lib/python3.10/site-packages (from bleach->nbconvert<7.0.0,>=6.1.0->datapane->polly-python) (0.5.1)\n",
            "Collecting wrapt<2,>=1.10\n",
            "  Downloading wrapt-1.14.1-cp310-cp310-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_17_x86_64.manylinux2014_x86_64.whl (77 kB)\n",
            "\u001b[K     |████████████████████████████████| 77 kB 15.9 MB/s  eta 0:00:01\n",
            "\u001b[?25hCollecting tenacity>=6.2.0\n",
            "  Downloading tenacity-8.1.0-py3-none-any.whl (23 kB)\n",
            "Collecting pluggy<2.0,>=0.12\n",
            "  Downloading pluggy-1.0.0-py2.py3-none-any.whl (13 kB)\n",
            "Collecting iniconfig\n",
            "  Downloading iniconfig-1.1.1-py2.py3-none-any.whl (5.0 kB)\n",
            "Collecting tomli>=1.0.0\n",
            "  Downloading tomli-2.0.1-py3-none-any.whl (12 kB)\n",
            "Collecting py>=1.8.2\n",
            "  Downloading py-1.11.0-py2.py3-none-any.whl (98 kB)\n",
            "\u001b[K     |████████████████████████████████| 98 kB 22.0 MB/s  eta 0:00:01\n",
            "\u001b[?25hRequirement already satisfied: docutils>=0.11 in /usr/local/lib/python3.10/site-packages (from rst2txt->polly-python) (0.15.2)\n",
            "Using legacy 'setup.py install' for Cerberus, since package 'wheel' is not installed.\n",
            "Using legacy 'setup.py install' for retrying, since package 'wheel' is not installed.\n",
            "Using legacy 'setup.py install' for micawber, since package 'wheel' is not installed.\n",
            "Using legacy 'setup.py install' for stringcase, since package 'wheel' is not installed.\n",
            "Using legacy 'setup.py install' for validators, since package 'wheel' is not installed.\n",
            "Installing collected packages: pytz, certifi, boltons, typing-extensions, toolz, requests, orderedmultidict, monotonic, face, backoff, wrapt, vega-datasets, validators, tomli, tenacity, stringcase, soupsieve, requests-toolbelt, python-magic, pydantic, pyarrow, py, posthog, pluggy, munch, micawber, lxml, iniconfig, importlib-resources, h5py, glom, furl, dulwich, dominate, datacommons-pandas, datacommons, dacite, colorlog, click-spinner, click, altair, sqlparse, rst2txt, retrying, pytest, postpy2, plotly, joblib, elucidatacloudpathlib, Deprecated, datapane, cmapPy, Cerberus, beautifulsoup4, polly-python\n",
            "  Attempting uninstall: pytz\n",
            "    Found existing installation: pytz 2021.3\n",
            "    Uninstalling pytz-2021.3:\n",
            "      Successfully uninstalled pytz-2021.3\n",
            "  Attempting uninstall: certifi\n",
            "    Found existing installation: certifi 2021.5.30\n",
            "    Uninstalling certifi-2021.5.30:\n",
            "      Successfully uninstalled certifi-2021.5.30\n",
            "  Attempting uninstall: requests\n",
            "    Found existing installation: requests 2.27.1\n",
            "    Uninstalling requests-2.27.1:\n",
            "      Successfully uninstalled requests-2.27.1\n",
            "    Running setup.py install for validators ... \u001b[?25ldone\n",
            "\u001b[?25h    Running setup.py install for stringcase ... \u001b[?25ldone\n",
            "\u001b[?25h    Running setup.py install for micawber ... \u001b[?25ldone\n",
            "\u001b[?25h    Running setup.py install for retrying ... \u001b[?25ldone\n",
            "\u001b[?25h  Attempting uninstall: Cerberus\n",
            "    Found existing installation: Cerberus 1.3.4\n",
            "    Uninstalling Cerberus-1.3.4:\n",
            "      Successfully uninstalled Cerberus-1.3.4\n",
            "    Running setup.py install for Cerberus ... \u001b[?25ldone\n",
            "\u001b[?25hSuccessfully installed Cerberus-1.3.2 Deprecated-1.2.13 altair-4.2.0 backoff-1.11.1 beautifulsoup4-4.11.1 boltons-21.0.0 certifi-2021.10.8 click-8.1.3 click-spinner-0.1.10 cmapPy-4.0.1 colorlog-6.7.0 dacite-1.6.0 datacommons-1.4.3 datacommons-pandas-0.0.3 datapane-0.15.3 dominate-2.7.0 dulwich-0.20.46 elucidatacloudpathlib-0.6.6 face-22.0.0 furl-2.1.3 glom-22.1.0 h5py-3.7.0 importlib-resources-5.10.0 iniconfig-1.1.1 joblib-1.2.0 lxml-4.9.1 micawber-0.5.4 monotonic-1.6 munch-2.5.0 orderedmultidict-1.0.1 plotly-5.10.0 pluggy-1.0.0 polly-python-0.2.2 posthog-2.1.2 postpy2-0.0.6 py-1.11.0 pyarrow-6.0.1 pydantic-1.9.1 pytest-7.1.3 python-magic-0.4.24 pytz-2021.1 requests-2.25.1 requests-toolbelt-0.9.1 retrying-1.3.3 rst2txt-1.1.0 soupsieve-2.3.2.post1 sqlparse-0.4.3 stringcase-1.2.0 tenacity-8.1.0 tomli-2.0.1 toolz-0.12.0 typing-extensions-4.4.0 validators-0.20.0 vega-datasets-0.9.0 wrapt-1.14.1\n",
            "\u001b[33mWARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv\u001b[0m\n"
          ]
        },
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "\u001b[33mWARNING: You are using pip version 21.2.4; however, version 22.3 is available.\r\n",
            "You should consider upgrading via the '/usr/local/bin/python3.10 -m pip install --upgrade pip' command.\u001b[0m\r\n"
          ]
        }
      ],
      "source": [
        "!sudo pip3 install polly-python"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "# Import Libraries and Authenticate"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 3,
      "metadata": {},
      "outputs": [],
      "source": [
        "import os\n",
        "from polly.auth import Polly\n",
        "from polly.omixatlas import OmixAtlas\n",
        "from polly.workspaces import Workspaces"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 4,
      "metadata": {},
      "outputs": [],
      "source": [
        "# AUTH_TOKEN=(os.environ['POLLY_REFRESH_TOKEN']) # Obtain authentication tokens\n",
        "AUTH_TOKEN = \"eyJjdHkiOiJKV1QiLCJlbmMiOiJBMjU2R0NNIiwiYWxnIjoiUlNBLU9BRVAifQ.OGp9MEwniWXDlkLSmDtdZTHY4fMF1X9fS3vyMmY0vvpk-5XCwoRUIuFtoOVksKBe7S32MsFa9o0TTMMudhEEVEeTZfBX_0ZiReLfnxqgQRIKP6HWCZtzEVPkos6iYcneFXjZBWDUQoD3cV1DSEZKsCkSwoe39sMSOjjNCHXUi4W1gMvl14lUYUCUmxvEhN8ob1cxrA8vBdIkobIYMrUkPEhtpdzo1_dEP7_LWYICdf4qNlzR0ki_VMGK4z-wlRGucD_T-OSYWxDHhnR7ba2ohFBlj6x4tOAqLa3SEtdkDltJhTshX8dvOohqkL3d3GI2KAbLY7MCJuyYENmO3MAPXA.HPE01XdRrna66gCc.d3719HXjYTqM5sWkfcej04MhNIDAM_SBh6Faq5-ELnvtuZlKhJqdrygduCNS9FWZl3zSyR9uUEE9jHD87YwPWsx3VPlfH4i2BC6wTi1Zoy9W30DE1NNBV3bNo7P4Xu6ljPYxbZ9E4gwM6wOfhU08wFrrEF6WxhGgnf4r4ILa6NWFn_yBVfBx6mPhqM1nv7cl1nXCqLJ7NgUSpi2UeC6KJf7SvlcNtpSTfXLtDjjBgYPpzMJQnDkhqcHead6nncpXTeRHtYuei3GaClX182BhpWN50kyFYVDa5xAZ8OWkrTgY2ZiaNE0Xf3pBK9vNj838IWagG2KBb3VqGN4IMYErHcSd73QtbJRZGpH5FUUFsk7D8L6iO0d_namwLzxceyAU9gOr0vY54gcT_UilrlG6dUzVJC5xyGriya-B7J0pyHxRFKQ1ApmC5c1LEAZUaJaNREX52MHRRO6o5CAHpMQpPB-HIh43y8ZP_CmtQSX1WM7zskcZJ6kLM1in_rMhsOzJojQcxhkBeyB0vlXTESTPRcME91FnYELXfasEnN7_g1yIy5OR-B_12B_ZCjpxxH1CZYuL8vNWulaDyDDPrLIzRcodvq9S_9uUby_yzG3TzGUS_jLUwzv65oA-cHbIJuxDLIYjOJRMuYqjEU7SKtmm57W08rO4v9wkZuQXuhB2-pZ3sKoJWwolMecxwiif3GLxfQjTSRm0aOvQl4nOn39JACwXAw5dtSSs-iNzQb3kcOvAu6V1LORZWgcehMoUDRfRbVqHP7P2SceA4zxRPfYPxrvX4GCOiT_FOTZfnbjCMJRlc7LggOdLg7dlXNYTuPOsyzm16F_Fr0HB5UfJy7Q80Sly5TJb15OKvriPZoVWmtNRaQZIaTgV32PB7mlfkFSJ0CUt5hlCMYdnnZ2EZN79Z1ZbY-NGH7fssp9paZVDPVIF5-KQpvvDSaBqmKMZko0jljNbLqtj0H_T7JRK2EU0cF6BvXPFJY0YBMb6sUnrmdREhLm5jF-bC_6g6n16l-hhgCXhfIejDdEorQNRHN87Tw_sqTl7QRrnIq4MASkOci1vGjl2o2N5tZFP3lcqp4b3m2Hl-cna29uAhviB5mi-5dQ2F6OXrNdooYy3FL_f-HlFsb8b3swwVaS6yTq71_P3KIf81TeMW1Nl2SyVzu8TqC5wJdHqScv5cbbb4ZsSfIAJmBf3wd7Zhj-KEc16_Jd-g3vEPpGGlqc9hGWKVs_EbyLfiY7h6LmWrmz3YwQSFKSFDNbZqPfV8c5r-KngjpNijCqNYScEkKJMSyugyCYh8wOSzoYeXrOQe1VDjkYprCzfvez0vfA07zzoFLYQ9IVVEqExpF--68oz.lAOoGssQp4oj39ouEU0Sug\"\n",
        "Polly.auth(AUTH_TOKEN, env = \"testpolly\")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 5,
      "metadata": {},
      "outputs": [],
      "source": [
        "pd.set_option('display.max_columns', 500)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "# Get Info and Schema of Omix Atlas"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 6,
      "metadata": {},
      "outputs": [],
      "source": [
        "# Defining omixatlas object\n",
        "omixatlas = OmixAtlas()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 7,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>repo_name</th>\n",
              "      <th>repo_id</th>\n",
              "      <th>indexes</th>\n",
              "      <th>v2_indexes</th>\n",
              "      <th>sources</th>\n",
              "      <th>datatypes</th>\n",
              "      <th>dataset_count</th>\n",
              "      <th>disease_count</th>\n",
              "      <th>tissue_count</th>\n",
              "      <th>organism_count</th>\n",
              "      <th>cell_line_count</th>\n",
              "      <th>cell_type_count</th>\n",
              "      <th>drug_count</th>\n",
              "      <th>data_type_count</th>\n",
              "      <th>data_source_count</th>\n",
              "      <th>sample_count</th>\n",
              "      <th>normal_sample_count</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>oa_testing</td>\n",
              "      <td>1659288377827</td>\n",
              "      <td>{}</td>\n",
              "      <td>{}</td>\n",
              "      <td>[]</td>\n",
              "      <td>[]</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>multiple_oa</td>\n",
              "      <td>1659420917030</td>\n",
              "      <td>{}</td>\n",
              "      <td>{}</td>\n",
              "      <td>[]</td>\n",
              "      <td>[]</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>ingestion_test_1</td>\n",
              "      <td>1654268055800</td>\n",
              "      <td>{}</td>\n",
              "      <td>{}</td>\n",
              "      <td>[]</td>\n",
              "      <td>[]</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>ingestion_demo</td>\n",
              "      <td>1656584298170</td>\n",
              "      <td>{}</td>\n",
              "      <td>{}</td>\n",
              "      <td>[]</td>\n",
              "      <td>[]</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>test_ingestion</td>\n",
              "      <td>1666170469120</td>\n",
              "      <td>{}</td>\n",
              "      <td>{}</td>\n",
              "      <td>[]</td>\n",
              "      <td>[]</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>cell_viability_assay</td>\n",
              "      <td>1649161733379</td>\n",
              "      <td>{'files': 'cell_viability_assay_files', 'gct_c...</td>\n",
              "      <td>{'files': 'cell_viability_assay_files', 'gct_c...</td>\n",
              "      <td>[]</td>\n",
              "      <td>[{'drug screening data': 571}]</td>\n",
              "      <td>571</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>lincs</td>\n",
              "      <td>31</td>\n",
              "      <td>{'files': 'lincs_files', 'gct_col_metadata': '...</td>\n",
              "      <td>{'files': 'lincs_files', 'gct_col_metadata': '...</td>\n",
              "      <td>[{'geo, lincs': 700}, {'geo': 95}]</td>\n",
              "      <td>[{'transcriptomics': 795}]</td>\n",
              "      <td>795</td>\n",
              "      <td>29</td>\n",
              "      <td>20</td>\n",
              "      <td>0</td>\n",
              "      <td>66</td>\n",
              "      <td>1</td>\n",
              "      <td>571</td>\n",
              "      <td>1</td>\n",
              "      <td>2</td>\n",
              "      <td>198200.0</td>\n",
              "      <td>9062</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>sc_data_lake</td>\n",
              "      <td>17</td>\n",
              "      <td>{'files': 'sc_data_lake_files', 'h5ad_col_meta...</td>\n",
              "      <td>{'files': 'sc_data_lake_files', 'h5ad_col_meta...</td>\n",
              "      <td>[{'geo': 32}, {'expression atlas': 1}, {'gene ...</td>\n",
              "      <td>[{'single cell': 35}]</td>\n",
              "      <td>36</td>\n",
              "      <td>28</td>\n",
              "      <td>34</td>\n",
              "      <td>3</td>\n",
              "      <td>2</td>\n",
              "      <td>53</td>\n",
              "      <td>5</td>\n",
              "      <td>1</td>\n",
              "      <td>5</td>\n",
              "      <td>4295.0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>8</th>\n",
              "      <td>liver_atlas</td>\n",
              "      <td>1622526550765</td>\n",
              "      <td>{'files': 'liver_atlas_files', 'gct_col_metada...</td>\n",
              "      <td>{'files': 'liver_atlas_files', 'gct_col_metada...</td>\n",
              "      <td>[{'geo': 295}, {'tcga': 75}, {'metabolights': ...</td>\n",
              "      <td>[{'transcriptomics': 305}, {'mutation': 101}, ...</td>\n",
              "      <td>490</td>\n",
              "      <td>231</td>\n",
              "      <td>78</td>\n",
              "      <td>11</td>\n",
              "      <td>63</td>\n",
              "      <td>82</td>\n",
              "      <td>3735</td>\n",
              "      <td>18</td>\n",
              "      <td>12</td>\n",
              "      <td>46190.0</td>\n",
              "      <td>15268</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9</th>\n",
              "      <td>cell_viability_omixatlas</td>\n",
              "      <td>1664182416962</td>\n",
              "      <td>{'files': 'cell_viability_omixatlas_files', 'g...</td>\n",
              "      <td>{'files': 'cell_viability_omixatlas_files', 'g...</td>\n",
              "      <td>[]</td>\n",
              "      <td>[{'drug response': 359}]</td>\n",
              "      <td>359</td>\n",
              "      <td>7</td>\n",
              "      <td>5</td>\n",
              "      <td>0</td>\n",
              "      <td>10</td>\n",
              "      <td>1</td>\n",
              "      <td>6</td>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>10</th>\n",
              "      <td>tcga</td>\n",
              "      <td>15</td>\n",
              "      <td>{'files': 'tcga_files'}</td>\n",
              "      <td>{'files': 'tcga_files'}</td>\n",
              "      <td>[{'tcga': 678}]</td>\n",
              "      <td>[{'transcriptomics': 156}, {'copy number varia...</td>\n",
              "      <td>678</td>\n",
              "      <td>34</td>\n",
              "      <td>55</td>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>101</td>\n",
              "      <td>6</td>\n",
              "      <td>1</td>\n",
              "      <td>678.0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>11</th>\n",
              "      <td>hmp_omixatlas</td>\n",
              "      <td>1659621145863</td>\n",
              "      <td>{'files': 'hmp_omixatlas_files', 'biom_col_met...</td>\n",
              "      <td>{'files': 'hmp_omixatlas_files', 'biom_col_met...</td>\n",
              "      <td>[{'human microbiome project (hmp)': 32}, {'mic...</td>\n",
              "      <td>[{'metagenomics': 26}, {'metatranscriptomics':...</td>\n",
              "      <td>33</td>\n",
              "      <td>1</td>\n",
              "      <td>2</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>16</td>\n",
              "      <td>3</td>\n",
              "      <td>2</td>\n",
              "      <td>1248.0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>12</th>\n",
              "      <td>c_oa</td>\n",
              "      <td>1663151941472</td>\n",
              "      <td>{'files': 'c_oa_files'}</td>\n",
              "      <td>{'files': 'c_oa_files'}</td>\n",
              "      <td>[{'': 10}]</td>\n",
              "      <td>[{'drug response': 10}]</td>\n",
              "      <td>10</td>\n",
              "      <td>3</td>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "      <td>5</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>10.0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>13</th>\n",
              "      <td>ai_in_drug_discovery_oa</td>\n",
              "      <td>1663919116821</td>\n",
              "      <td>{'files': 'ai_in_drug_discovery_oa_files', 'gc...</td>\n",
              "      <td>{'files': 'ai_in_drug_discovery_oa_files', 'gc...</td>\n",
              "      <td>[{'geo': 129}]</td>\n",
              "      <td>[{'transcriptomics': 129}]</td>\n",
              "      <td>129</td>\n",
              "      <td>68</td>\n",
              "      <td>43</td>\n",
              "      <td>6</td>\n",
              "      <td>28</td>\n",
              "      <td>61</td>\n",
              "      <td>55</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>3649.0</td>\n",
              "      <td>1574</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>14</th>\n",
              "      <td>geo</td>\n",
              "      <td>1643016586529</td>\n",
              "      <td>{'files': 'geo_files', 'gct_col_metadata': 'ge...</td>\n",
              "      <td>{'files': 'geo_files', 'gct_col_metadata': 'ge...</td>\n",
              "      <td>[{'geo': 9994}]</td>\n",
              "      <td>[{'transcriptomics': 9991}, {'raw counts trans...</td>\n",
              "      <td>9994</td>\n",
              "      <td>1652</td>\n",
              "      <td>622</td>\n",
              "      <td>26</td>\n",
              "      <td>1918</td>\n",
              "      <td>525</td>\n",
              "      <td>560</td>\n",
              "      <td>2</td>\n",
              "      <td>1</td>\n",
              "      <td>233480.0</td>\n",
              "      <td>152883</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>15</th>\n",
              "      <td>p_oa</td>\n",
              "      <td>1662102491212</td>\n",
              "      <td>{'files': 'p_oa_files', 'gct_col_metadata': 'p...</td>\n",
              "      <td>{'files': 'p_oa_files', 'gct_col_metadata': 'p...</td>\n",
              "      <td>[{'tcga': 35}, {'geo': 20}]</td>\n",
              "      <td>[{'transcriptomics': 40}, {'copy number variat...</td>\n",
              "      <td>55</td>\n",
              "      <td>12</td>\n",
              "      <td>10</td>\n",
              "      <td>3</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>9</td>\n",
              "      <td>2</td>\n",
              "      <td>2</td>\n",
              "      <td>138.0</td>\n",
              "      <td>73</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>16</th>\n",
              "      <td>depmap_omixatlas</td>\n",
              "      <td>1647941576494</td>\n",
              "      <td>{'files': 'depmap_omixatlas_files', 'gct_col_m...</td>\n",
              "      <td>{'files': 'depmap_omixatlas_files', 'gct_col_m...</td>\n",
              "      <td>[{'depmap': 546}]</td>\n",
              "      <td>[{'gene dependency': 487}, {'drug screens': 59}]</td>\n",
              "      <td>546</td>\n",
              "      <td>89</td>\n",
              "      <td>37</td>\n",
              "      <td>1</td>\n",
              "      <td>410</td>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "      <td>2</td>\n",
              "      <td>1</td>\n",
              "      <td>546.0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>17</th>\n",
              "      <td>exelixis</td>\n",
              "      <td>1655454542384</td>\n",
              "      <td>{'files': 'exelixis_files', 'gct_col_metadata'...</td>\n",
              "      <td>{'files': 'exelixis_files', 'gct_col_metadata'...</td>\n",
              "      <td>[{'geo': 11}]</td>\n",
              "      <td>[{'transcriptomics': 11}]</td>\n",
              "      <td>11</td>\n",
              "      <td>4</td>\n",
              "      <td>3</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>3</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>1039.0</td>\n",
              "      <td>160</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>18</th>\n",
              "      <td>multiple_oa_2</td>\n",
              "      <td>1659450268526</td>\n",
              "      <td>{'files': 'multiple_oa_2_files', 'gct_col_meta...</td>\n",
              "      <td>{'files': 'multiple_oa_2_files', 'gct_col_meta...</td>\n",
              "      <td>[{'cbioportal': 226}, {'geo': 165}, {'depmap':...</td>\n",
              "      <td>[{'copy number variation': 192}, {'transcripto...</td>\n",
              "      <td>406</td>\n",
              "      <td>161</td>\n",
              "      <td>127</td>\n",
              "      <td>7</td>\n",
              "      <td>159</td>\n",
              "      <td>62</td>\n",
              "      <td>71</td>\n",
              "      <td>8</td>\n",
              "      <td>6</td>\n",
              "      <td>276262.0</td>\n",
              "      <td>183062</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>19</th>\n",
              "      <td>oye_vm_vm_oa</td>\n",
              "      <td>1661250513943</td>\n",
              "      <td>{'files': 'oye_vm_vm_oa_files', 'gct_col_metad...</td>\n",
              "      <td>{'files': 'oye_vm_vm_oa_files', 'gct_col_metad...</td>\n",
              "      <td>[{'pride': 4}]</td>\n",
              "      <td>[{'proteomics': 4}]</td>\n",
              "      <td>4</td>\n",
              "      <td>4</td>\n",
              "      <td>5</td>\n",
              "      <td>2</td>\n",
              "      <td>2</td>\n",
              "      <td>2</td>\n",
              "      <td>3</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>0.0</td>\n",
              "      <td>308</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>20</th>\n",
              "      <td>ingestion_testing_oa</td>\n",
              "      <td>1657777478388</td>\n",
              "      <td>{'files': 'ingestion_testing_oa_files', 'gct_c...</td>\n",
              "      <td>{'files': 'ingestion_testing_oa_files', 'gct_c...</td>\n",
              "      <td>[{'pride': 1}]</td>\n",
              "      <td>[{'proteomics': 1}]</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>2</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>3.0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>21</th>\n",
              "      <td>gnomad</td>\n",
              "      <td>1629715166552</td>\n",
              "      <td>{'files': 'gnomad_files'}</td>\n",
              "      <td>{'files': 'gnomad_files'}</td>\n",
              "      <td>[]</td>\n",
              "      <td>[]</td>\n",
              "      <td>720</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>22</th>\n",
              "      <td>clinical_oa_2</td>\n",
              "      <td>1664439952209</td>\n",
              "      <td>{'files': 'clinical_oa_2_files'}</td>\n",
              "      <td>{'files': 'clinical_oa_2_files'}</td>\n",
              "      <td>[{'teikobio': 16}]</td>\n",
              "      <td>[{'flow cytometry': 16}]</td>\n",
              "      <td>16</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>16.0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>23</th>\n",
              "      <td>ukbiobank</td>\n",
              "      <td>1644400393649</td>\n",
              "      <td>{'files': 'ukbiobank_files'}</td>\n",
              "      <td>{'files': 'ukbiobank_files'}</td>\n",
              "      <td>[]</td>\n",
              "      <td>[{'gwas': 648}]</td>\n",
              "      <td>648</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>24</th>\n",
              "      <td>testing_ingestion_bhavya</td>\n",
              "      <td>1659008683314</td>\n",
              "      <td>{'files': 'testing_ingestion_bhavya_files', 'g...</td>\n",
              "      <td>{'files': 'testing_ingestion_bhavya_files', 'g...</td>\n",
              "      <td>[{'geo': 4}]</td>\n",
              "      <td>[{'transcriptomics': 4}]</td>\n",
              "      <td>4</td>\n",
              "      <td>4</td>\n",
              "      <td>4</td>\n",
              "      <td>1</td>\n",
              "      <td>2</td>\n",
              "      <td>3</td>\n",
              "      <td>3</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>122.0</td>\n",
              "      <td>29</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                   repo_name        repo_id  \\\n",
              "0                 oa_testing  1659288377827   \n",
              "1                multiple_oa  1659420917030   \n",
              "2           ingestion_test_1  1654268055800   \n",
              "3             ingestion_demo  1656584298170   \n",
              "4             test_ingestion  1666170469120   \n",
              "5       cell_viability_assay  1649161733379   \n",
              "6                      lincs             31   \n",
              "7               sc_data_lake             17   \n",
              "8                liver_atlas  1622526550765   \n",
              "9   cell_viability_omixatlas  1664182416962   \n",
              "10                      tcga             15   \n",
              "11             hmp_omixatlas  1659621145863   \n",
              "12                      c_oa  1663151941472   \n",
              "13   ai_in_drug_discovery_oa  1663919116821   \n",
              "14                       geo  1643016586529   \n",
              "15                      p_oa  1662102491212   \n",
              "16          depmap_omixatlas  1647941576494   \n",
              "17                  exelixis  1655454542384   \n",
              "18             multiple_oa_2  1659450268526   \n",
              "19              oye_vm_vm_oa  1661250513943   \n",
              "20      ingestion_testing_oa  1657777478388   \n",
              "21                    gnomad  1629715166552   \n",
              "22             clinical_oa_2  1664439952209   \n",
              "23                 ukbiobank  1644400393649   \n",
              "24  testing_ingestion_bhavya  1659008683314   \n",
              "\n",
              "                                              indexes  \\\n",
              "0                                                  {}   \n",
              "1                                                  {}   \n",
              "2                                                  {}   \n",
              "3                                                  {}   \n",
              "4                                                  {}   \n",
              "5   {'files': 'cell_viability_assay_files', 'gct_c...   \n",
              "6   {'files': 'lincs_files', 'gct_col_metadata': '...   \n",
              "7   {'files': 'sc_data_lake_files', 'h5ad_col_meta...   \n",
              "8   {'files': 'liver_atlas_files', 'gct_col_metada...   \n",
              "9   {'files': 'cell_viability_omixatlas_files', 'g...   \n",
              "10                            {'files': 'tcga_files'}   \n",
              "11  {'files': 'hmp_omixatlas_files', 'biom_col_met...   \n",
              "12                            {'files': 'c_oa_files'}   \n",
              "13  {'files': 'ai_in_drug_discovery_oa_files', 'gc...   \n",
              "14  {'files': 'geo_files', 'gct_col_metadata': 'ge...   \n",
              "15  {'files': 'p_oa_files', 'gct_col_metadata': 'p...   \n",
              "16  {'files': 'depmap_omixatlas_files', 'gct_col_m...   \n",
              "17  {'files': 'exelixis_files', 'gct_col_metadata'...   \n",
              "18  {'files': 'multiple_oa_2_files', 'gct_col_meta...   \n",
              "19  {'files': 'oye_vm_vm_oa_files', 'gct_col_metad...   \n",
              "20  {'files': 'ingestion_testing_oa_files', 'gct_c...   \n",
              "21                          {'files': 'gnomad_files'}   \n",
              "22                   {'files': 'clinical_oa_2_files'}   \n",
              "23                       {'files': 'ukbiobank_files'}   \n",
              "24  {'files': 'testing_ingestion_bhavya_files', 'g...   \n",
              "\n",
              "                                           v2_indexes  \\\n",
              "0                                                  {}   \n",
              "1                                                  {}   \n",
              "2                                                  {}   \n",
              "3                                                  {}   \n",
              "4                                                  {}   \n",
              "5   {'files': 'cell_viability_assay_files', 'gct_c...   \n",
              "6   {'files': 'lincs_files', 'gct_col_metadata': '...   \n",
              "7   {'files': 'sc_data_lake_files', 'h5ad_col_meta...   \n",
              "8   {'files': 'liver_atlas_files', 'gct_col_metada...   \n",
              "9   {'files': 'cell_viability_omixatlas_files', 'g...   \n",
              "10                            {'files': 'tcga_files'}   \n",
              "11  {'files': 'hmp_omixatlas_files', 'biom_col_met...   \n",
              "12                            {'files': 'c_oa_files'}   \n",
              "13  {'files': 'ai_in_drug_discovery_oa_files', 'gc...   \n",
              "14  {'files': 'geo_files', 'gct_col_metadata': 'ge...   \n",
              "15  {'files': 'p_oa_files', 'gct_col_metadata': 'p...   \n",
              "16  {'files': 'depmap_omixatlas_files', 'gct_col_m...   \n",
              "17  {'files': 'exelixis_files', 'gct_col_metadata'...   \n",
              "18  {'files': 'multiple_oa_2_files', 'gct_col_meta...   \n",
              "19  {'files': 'oye_vm_vm_oa_files', 'gct_col_metad...   \n",
              "20  {'files': 'ingestion_testing_oa_files', 'gct_c...   \n",
              "21                          {'files': 'gnomad_files'}   \n",
              "22                   {'files': 'clinical_oa_2_files'}   \n",
              "23                       {'files': 'ukbiobank_files'}   \n",
              "24  {'files': 'testing_ingestion_bhavya_files', 'g...   \n",
              "\n",
              "                                              sources  \\\n",
              "0                                                  []   \n",
              "1                                                  []   \n",
              "2                                                  []   \n",
              "3                                                  []   \n",
              "4                                                  []   \n",
              "5                                                  []   \n",
              "6                  [{'geo, lincs': 700}, {'geo': 95}]   \n",
              "7   [{'geo': 32}, {'expression atlas': 1}, {'gene ...   \n",
              "8   [{'geo': 295}, {'tcga': 75}, {'metabolights': ...   \n",
              "9                                                  []   \n",
              "10                                    [{'tcga': 678}]   \n",
              "11  [{'human microbiome project (hmp)': 32}, {'mic...   \n",
              "12                                         [{'': 10}]   \n",
              "13                                     [{'geo': 129}]   \n",
              "14                                    [{'geo': 9994}]   \n",
              "15                        [{'tcga': 35}, {'geo': 20}]   \n",
              "16                                  [{'depmap': 546}]   \n",
              "17                                      [{'geo': 11}]   \n",
              "18  [{'cbioportal': 226}, {'geo': 165}, {'depmap':...   \n",
              "19                                     [{'pride': 4}]   \n",
              "20                                     [{'pride': 1}]   \n",
              "21                                                 []   \n",
              "22                                 [{'teikobio': 16}]   \n",
              "23                                                 []   \n",
              "24                                       [{'geo': 4}]   \n",
              "\n",
              "                                            datatypes  dataset_count  \\\n",
              "0                                                  []              0   \n",
              "1                                                  []              0   \n",
              "2                                                  []              0   \n",
              "3                                                  []              0   \n",
              "4                                                  []              0   \n",
              "5                      [{'drug screening data': 571}]            571   \n",
              "6                          [{'transcriptomics': 795}]            795   \n",
              "7                               [{'single cell': 35}]             36   \n",
              "8   [{'transcriptomics': 305}, {'mutation': 101}, ...            490   \n",
              "9                            [{'drug response': 359}]            359   \n",
              "10  [{'transcriptomics': 156}, {'copy number varia...            678   \n",
              "11  [{'metagenomics': 26}, {'metatranscriptomics':...             33   \n",
              "12                            [{'drug response': 10}]             10   \n",
              "13                         [{'transcriptomics': 129}]            129   \n",
              "14  [{'transcriptomics': 9991}, {'raw counts trans...           9994   \n",
              "15  [{'transcriptomics': 40}, {'copy number variat...             55   \n",
              "16   [{'gene dependency': 487}, {'drug screens': 59}]            546   \n",
              "17                          [{'transcriptomics': 11}]             11   \n",
              "18  [{'copy number variation': 192}, {'transcripto...            406   \n",
              "19                                [{'proteomics': 4}]              4   \n",
              "20                                [{'proteomics': 1}]              1   \n",
              "21                                                 []            720   \n",
              "22                           [{'flow cytometry': 16}]             16   \n",
              "23                                    [{'gwas': 648}]            648   \n",
              "24                           [{'transcriptomics': 4}]              4   \n",
              "\n",
              "    disease_count  tissue_count  organism_count  cell_line_count  \\\n",
              "0               0             0               0                0   \n",
              "1               0             0               0                0   \n",
              "2               0             0               0                0   \n",
              "3               0             0               0                0   \n",
              "4               0             0               0                0   \n",
              "5               0             0               0                0   \n",
              "6              29            20               0               66   \n",
              "7              28            34               3                2   \n",
              "8             231            78              11               63   \n",
              "9               7             5               0               10   \n",
              "10             34            55               1                0   \n",
              "11              1             2               1                1   \n",
              "12              3             1               0                5   \n",
              "13             68            43               6               28   \n",
              "14           1652           622              26             1918   \n",
              "15             12            10               3                0   \n",
              "16             89            37               1              410   \n",
              "17              4             3               1                1   \n",
              "18            161           127               7              159   \n",
              "19              4             5               2                2   \n",
              "20              1             1               1                1   \n",
              "21              0             0               1                0   \n",
              "22              1             1               1                0   \n",
              "23              0             0               1                0   \n",
              "24              4             4               1                2   \n",
              "\n",
              "    cell_type_count  drug_count  data_type_count  data_source_count  \\\n",
              "0                 0           0                0                  0   \n",
              "1                 0           0                0                  0   \n",
              "2                 0           0                0                  0   \n",
              "3                 0           0                0                  0   \n",
              "4                 0           0                0                  0   \n",
              "5                 0           0                1                  0   \n",
              "6                 1         571                1                  2   \n",
              "7                53           5                1                  5   \n",
              "8                82        3735               18                 12   \n",
              "9                 1           6                1                  0   \n",
              "10                0         101                6                  1   \n",
              "11                1          16                3                  2   \n",
              "12                1           1                1                  1   \n",
              "13               61          55                1                  1   \n",
              "14              525         560                2                  1   \n",
              "15                0           9                2                  2   \n",
              "16                1           0                2                  1   \n",
              "17                1           3                1                  1   \n",
              "18               62          71                8                  6   \n",
              "19                2           3                1                  1   \n",
              "20                1           2                1                  1   \n",
              "21                0           0                0                  0   \n",
              "22                1           0                1                  1   \n",
              "23                0           0                1                  0   \n",
              "24                3           3                1                  1   \n",
              "\n",
              "    sample_count  normal_sample_count  \n",
              "0            0.0                    0  \n",
              "1            0.0                    0  \n",
              "2            0.0                    0  \n",
              "3            0.0                    0  \n",
              "4            0.0                    0  \n",
              "5            0.0                    0  \n",
              "6       198200.0                 9062  \n",
              "7         4295.0                    0  \n",
              "8        46190.0                15268  \n",
              "9            0.0                    0  \n",
              "10         678.0                    0  \n",
              "11        1248.0                    0  \n",
              "12          10.0                    0  \n",
              "13        3649.0                 1574  \n",
              "14      233480.0               152883  \n",
              "15         138.0                   73  \n",
              "16         546.0                    0  \n",
              "17        1039.0                  160  \n",
              "18      276262.0               183062  \n",
              "19           0.0                  308  \n",
              "20           3.0                    0  \n",
              "21           0.0                    0  \n",
              "22          16.0                    0  \n",
              "23           0.0                    0  \n",
              "24         122.0                   29  "
            ]
          },
          "execution_count": 7,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Get all omix atlases available to user\n",
        "all_omix_atlas = pd.DataFrame.from_dict(omixatlas.get_all_omixatlas()['data'])\n",
        "all_omix_atlas"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 8,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "{'data': {'repo_name': 'multiple_oa_2',\n",
              "  'repo_id': '1659450268526',\n",
              "  'indexes': {'files': 'multiple_oa_2_files',\n",
              "   'gct_col_metadata': 'multiple_oa_2_gct_col_metadata'},\n",
              "  'v2_indexes': {'files': 'multiple_oa_2_files',\n",
              "   'gct_col_metadata': 'multiple_oa_2_gct_col_metadata'},\n",
              "  'sources': [{'cbioportal': 226},\n",
              "   {'geo': 165},\n",
              "   {'depmap': 5},\n",
              "   {'tyra': 5},\n",
              "   {'ccle': 4},\n",
              "   {'cptac': 1}],\n",
              "  'datatypes': [{'copy number variation': 192},\n",
              "   {'transcriptomics': 165},\n",
              "   {'mutation': 33},\n",
              "   {'fusion': 4},\n",
              "   {'gene dependency': 4},\n",
              "   {'drug screens': 1},\n",
              "   {'metabolomics': 1},\n",
              "   {'mirna': 1}],\n",
              "  'dataset_count': 406,\n",
              "  'disease_count': 161,\n",
              "  'tissue_count': 127,\n",
              "  'organism_count': 7,\n",
              "  'cell_line_count': 159,\n",
              "  'cell_type_count': 62,\n",
              "  'drug_count': 71,\n",
              "  'data_type_count': 8,\n",
              "  'data_source_count': 6,\n",
              "  'sample_count': 276262.0,\n",
              "  'normal_sample_count': 183062}}"
            ]
          },
          "execution_count": 8,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Get summary of selected Omix Atlas - multiple_oa_2\n",
        "omix_atlas_summary = omixatlas.omixatlas_summary('1659450268526')\n",
        "# OR USE REPO ID- omix_atlas_summary = pd.DataFrame.from_dict(omixatlas.omixatlas_summary('1615965444377'))['data']\n",
        "omix_atlas_summary"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 9,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Query execution succeeded (time taken: 0.88 seconds, data scanned: 0.000 MB)\n",
            "Fetched 2 rows\n"
          ]
        },
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>table_name</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>multiple_oa_2.datasets</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>multiple_oa_2.samples</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "               table_name\n",
              "0  multiple_oa_2.datasets\n",
              "1   multiple_oa_2.samples"
            ]
          },
          "execution_count": 9,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Query for tables in multiple_oa_2\n",
        "query = \"SHOW TABLES IN multiple_oa_2\"\n",
        "results=omixatlas.query_metadata(query)\n",
        "results"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 10,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Query execution succeeded (time taken: 1.92 seconds, data scanned: 0.000 MB)\n",
            "Fetched 48 rows\n"
          ]
        },
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>column_name</th>\n",
              "      <th>column_type</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>data_matrix_available</td>\n",
              "      <td>string</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>curated_organism</td>\n",
              "      <td>array&lt;string&gt;</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>src_uri</td>\n",
              "      <td>string</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>total_num_samples</td>\n",
              "      <td>bigint</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>year</td>\n",
              "      <td>string</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>description</td>\n",
              "      <td>string</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>curated_cell_line</td>\n",
              "      <td>array&lt;string&gt;</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>data_table_name</td>\n",
              "      <td>string</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>8</th>\n",
              "      <td>data_table_version</td>\n",
              "      <td>bigint</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9</th>\n",
              "      <td>platform</td>\n",
              "      <td>string</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>10</th>\n",
              "      <td>timestamp_</td>\n",
              "      <td>string</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>11</th>\n",
              "      <td>subseries_ids</td>\n",
              "      <td>array&lt;string&gt;</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>12</th>\n",
              "      <td>file_type</td>\n",
              "      <td>string</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>13</th>\n",
              "      <td>publication</td>\n",
              "      <td>string</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>14</th>\n",
              "      <td>missing_samples</td>\n",
              "      <td>array&lt;string&gt;</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>15</th>\n",
              "      <td>curated_cell_type</td>\n",
              "      <td>array&lt;string&gt;</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>16</th>\n",
              "      <td>key</td>\n",
              "      <td>string</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>17</th>\n",
              "      <td>summary</td>\n",
              "      <td>string</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>18</th>\n",
              "      <td>src_repo</td>\n",
              "      <td>string</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>19</th>\n",
              "      <td>drug_smiles</td>\n",
              "      <td>array&lt;string&gt;</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>20</th>\n",
              "      <td>package</td>\n",
              "      <td>string</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>21</th>\n",
              "      <td>file_location</td>\n",
              "      <td>string</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>22</th>\n",
              "      <td>author</td>\n",
              "      <td>string</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>23</th>\n",
              "      <td>dataset_id</td>\n",
              "      <td>string</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>24</th>\n",
              "      <td>curated_disease</td>\n",
              "      <td>array&lt;string&gt;</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>25</th>\n",
              "      <td>curated_drug</td>\n",
              "      <td>array&lt;string&gt;</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>26</th>\n",
              "      <td>curated_gene</td>\n",
              "      <td>array&lt;string&gt;</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>27</th>\n",
              "      <td>abstract</td>\n",
              "      <td>string</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>28</th>\n",
              "      <td>version</td>\n",
              "      <td>bigint</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>29</th>\n",
              "      <td>curated_strain</td>\n",
              "      <td>array&lt;string&gt;</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>30</th>\n",
              "      <td>bucket</td>\n",
              "      <td>string</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>31</th>\n",
              "      <td>curated_tissue</td>\n",
              "      <td>array&lt;string&gt;</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>32</th>\n",
              "      <td>dataset_source</td>\n",
              "      <td>string</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>33</th>\n",
              "      <td>data_type</td>\n",
              "      <td>string</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>34</th>\n",
              "      <td>overall_design</td>\n",
              "      <td>string</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>35</th>\n",
              "      <td>is_current</td>\n",
              "      <td>string</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>36</th>\n",
              "      <td>region</td>\n",
              "      <td>string</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>37</th>\n",
              "      <td>curated_file_type</td>\n",
              "      <td>string</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>38</th>\n",
              "      <td>curated_disease_type</td>\n",
              "      <td>string</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>39</th>\n",
              "      <td>sample_id</td>\n",
              "      <td>string</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>40</th>\n",
              "      <td>raw_disease</td>\n",
              "      <td>array&lt;string&gt;</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>41</th>\n",
              "      <td>curation_version</td>\n",
              "      <td>string</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>42</th>\n",
              "      <td>data_repository</td>\n",
              "      <td>string</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>43</th>\n",
              "      <td>experimental_design</td>\n",
              "      <td>string</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>44</th>\n",
              "      <td>data_source</td>\n",
              "      <td>string</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>45</th>\n",
              "      <td>curated_repo</td>\n",
              "      <td>string</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>46</th>\n",
              "      <td>curated_project</td>\n",
              "      <td>string</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>47</th>\n",
              "      <td>operation</td>\n",
              "      <td>string</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "              column_name    column_type\n",
              "0   data_matrix_available         string\n",
              "1        curated_organism  array<string>\n",
              "2                 src_uri         string\n",
              "3       total_num_samples         bigint\n",
              "4                    year         string\n",
              "5             description         string\n",
              "6       curated_cell_line  array<string>\n",
              "7         data_table_name         string\n",
              "8      data_table_version         bigint\n",
              "9                platform         string\n",
              "10             timestamp_         string\n",
              "11          subseries_ids  array<string>\n",
              "12              file_type         string\n",
              "13            publication         string\n",
              "14        missing_samples  array<string>\n",
              "15      curated_cell_type  array<string>\n",
              "16                    key         string\n",
              "17                summary         string\n",
              "18               src_repo         string\n",
              "19            drug_smiles  array<string>\n",
              "20                package         string\n",
              "21          file_location         string\n",
              "22                 author         string\n",
              "23             dataset_id         string\n",
              "24        curated_disease  array<string>\n",
              "25           curated_drug  array<string>\n",
              "26           curated_gene  array<string>\n",
              "27               abstract         string\n",
              "28                version         bigint\n",
              "29         curated_strain  array<string>\n",
              "30                 bucket         string\n",
              "31         curated_tissue  array<string>\n",
              "32         dataset_source         string\n",
              "33              data_type         string\n",
              "34         overall_design         string\n",
              "35             is_current         string\n",
              "36                 region         string\n",
              "37      curated_file_type         string\n",
              "38   curated_disease_type         string\n",
              "39              sample_id         string\n",
              "40            raw_disease  array<string>\n",
              "41       curation_version         string\n",
              "42        data_repository         string\n",
              "43    experimental_design         string\n",
              "44            data_source         string\n",
              "45           curated_repo         string\n",
              "46        curated_project         string\n",
              "47              operation         string"
            ]
          },
          "execution_count": 10,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# You can run the same for samples as well\n",
        "query = \"DESCRIBE multiple_oa_2.datasets\"\n",
        "results=omixatlas.query_metadata(query)\n",
        "results"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 11,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Source</th>\n",
              "      <th>Datatype</th>\n",
              "      <th>Field Name</th>\n",
              "      <th>Field Description</th>\n",
              "      <th>Field Type</th>\n",
              "      <th>Is Curated</th>\n",
              "      <th>Is Array</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>geo</td>\n",
              "      <td>all</td>\n",
              "      <td>abstract</td>\n",
              "      <td>Abstract of the publication assocaited with th...</td>\n",
              "      <td>text</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>geo</td>\n",
              "      <td>all</td>\n",
              "      <td>author</td>\n",
              "      <td>Name of the author who published the dataset</td>\n",
              "      <td>text</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>geo</td>\n",
              "      <td>all</td>\n",
              "      <td>bucket</td>\n",
              "      <td>S3 bucket in which the data file resides</td>\n",
              "      <td>text</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>cbioportal</td>\n",
              "      <td>all</td>\n",
              "      <td>bucket</td>\n",
              "      <td>S3 bucket in which the data file resides</td>\n",
              "      <td>text</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>cbioportal</td>\n",
              "      <td>all</td>\n",
              "      <td>curated_cell_line</td>\n",
              "      <td>Cell lines from which the samples were derived...</td>\n",
              "      <td>text</td>\n",
              "      <td>False</td>\n",
              "      <td>True</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>68</th>\n",
              "      <td>cbioportal</td>\n",
              "      <td>all</td>\n",
              "      <td>total_num_samples</td>\n",
              "      <td>Total number of samples in a dataset</td>\n",
              "      <td>integer</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>69</th>\n",
              "      <td>geo</td>\n",
              "      <td>all</td>\n",
              "      <td>total_num_samples</td>\n",
              "      <td>Total number of samples in a dataset</td>\n",
              "      <td>integer</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>70</th>\n",
              "      <td>geo</td>\n",
              "      <td>all</td>\n",
              "      <td>version</td>\n",
              "      <td>Version of data file</td>\n",
              "      <td>integer</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>71</th>\n",
              "      <td>cbioportal</td>\n",
              "      <td>all</td>\n",
              "      <td>version</td>\n",
              "      <td>Version of data file</td>\n",
              "      <td>integer</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>72</th>\n",
              "      <td>geo</td>\n",
              "      <td>all</td>\n",
              "      <td>year</td>\n",
              "      <td>Year in which the dataset was published</td>\n",
              "      <td>text</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>73 rows × 7 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "        Source Datatype         Field Name  \\\n",
              "0          geo      all           abstract   \n",
              "1          geo      all             author   \n",
              "2          geo      all             bucket   \n",
              "3   cbioportal      all             bucket   \n",
              "4   cbioportal      all  curated_cell_line   \n",
              "..         ...      ...                ...   \n",
              "68  cbioportal      all  total_num_samples   \n",
              "69         geo      all  total_num_samples   \n",
              "70         geo      all            version   \n",
              "71  cbioportal      all            version   \n",
              "72         geo      all               year   \n",
              "\n",
              "                                    Field Description Field Type  Is Curated  \\\n",
              "0   Abstract of the publication assocaited with th...       text       False   \n",
              "1        Name of the author who published the dataset       text       False   \n",
              "2            S3 bucket in which the data file resides       text       False   \n",
              "3            S3 bucket in which the data file resides       text       False   \n",
              "4   Cell lines from which the samples were derived...       text       False   \n",
              "..                                                ...        ...         ...   \n",
              "68               Total number of samples in a dataset    integer       False   \n",
              "69               Total number of samples in a dataset    integer       False   \n",
              "70                               Version of data file    integer       False   \n",
              "71                               Version of data file    integer       False   \n",
              "72            Year in which the dataset was published       text       False   \n",
              "\n",
              "    Is Array  \n",
              "0      False  \n",
              "1      False  \n",
              "2      False  \n",
              "3      False  \n",
              "4       True  \n",
              "..       ...  \n",
              "68     False  \n",
              "69     False  \n",
              "70     False  \n",
              "71     False  \n",
              "72     False  \n",
              "\n",
              "[73 rows x 7 columns]"
            ]
          },
          "execution_count": 11,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Get schema for multiple_oa_2. Use parameter return_type='dict' if you want o/p as dictionary\n",
        "schema = omixatlas.get_schema(1659450268526)\n",
        "schema.datasets"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "# Query for multiple sources and datatypes"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "Query structure for multiple source: \n",
        "-> query = \"\"\"SELECT * FROM repo_name.source_name_in_schema.datasets \n",
        "        WHERE CONTAINS(curated_disease, 'Multiple Myeloma') \"\"\"\n",
        "        \n",
        "Query structure for multiple source:\n",
        "-> query = \"\"\"SELECT * FROM repo_name.source_name_in_schema.datatype_name_in_schema.datasets \n",
        "        WHERE CONTAINS(curated_disease, 'Multiple Myeloma') \"\"\""
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 12,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Query execution succeeded (time taken: 2.13 seconds, data scanned: 0.092 MB)\n",
            "Fetched 408 rows\n"
          ]
        },
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>dataset_id</th>\n",
              "      <th>abstract</th>\n",
              "      <th>author</th>\n",
              "      <th>bucket</th>\n",
              "      <th>curated_cell_line</th>\n",
              "      <th>curated_cell_type</th>\n",
              "      <th>curated_disease</th>\n",
              "      <th>curated_drug</th>\n",
              "      <th>curated_gene</th>\n",
              "      <th>curated_organism</th>\n",
              "      <th>curated_strain</th>\n",
              "      <th>curated_tissue</th>\n",
              "      <th>data_matrix_available</th>\n",
              "      <th>data_table_name</th>\n",
              "      <th>data_table_version</th>\n",
              "      <th>data_type</th>\n",
              "      <th>dataset_source</th>\n",
              "      <th>description</th>\n",
              "      <th>drug_smiles</th>\n",
              "      <th>file_location</th>\n",
              "      <th>file_type</th>\n",
              "      <th>is_current</th>\n",
              "      <th>key</th>\n",
              "      <th>overall_design</th>\n",
              "      <th>package</th>\n",
              "      <th>platform</th>\n",
              "      <th>publication</th>\n",
              "      <th>region</th>\n",
              "      <th>src_repo</th>\n",
              "      <th>src_uri</th>\n",
              "      <th>summary</th>\n",
              "      <th>timestamp_</th>\n",
              "      <th>total_num_samples</th>\n",
              "      <th>version</th>\n",
              "      <th>year</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>01CO001_cnv</td>\n",
              "      <td></td>\n",
              "      <td></td>\n",
              "      <td>discover-test-datalake-v1</td>\n",
              "      <td>[None]</td>\n",
              "      <td>[None]</td>\n",
              "      <td>[Colonic Neoplasms]</td>\n",
              "      <td>[None]</td>\n",
              "      <td>[]</td>\n",
              "      <td>[Homo Sapiens]</td>\n",
              "      <td>[]</td>\n",
              "      <td>[colon sigmoideum]</td>\n",
              "      <td></td>\n",
              "      <td>multiple_oa_2__01co001_cnv</td>\n",
              "      <td>0</td>\n",
              "      <td>['copy number variation']</td>\n",
              "      <td>CPTAC</td>\n",
              "      <td>Copy number variation values for 01CO001 in co...</td>\n",
              "      <td>[]</td>\n",
              "      <td>https://discover-test-datalake-v1.s3-ap-southe...</td>\n",
              "      <td>gct</td>\n",
              "      <td>true</td>\n",
              "      <td>multiple_oa_2/data/220829-1818/01CO001_cnv.gct</td>\n",
              "      <td></td>\n",
              "      <td>multiple_oa_2/data</td>\n",
              "      <td></td>\n",
              "      <td>https://doi.org/10.1016/j.cell.2019.03.030</td>\n",
              "      <td>ap-southeast-1</td>\n",
              "      <td>multiple_oa_2</td>\n",
              "      <td>polly:data://multiple_oa_2/data/220829-1818/01...</td>\n",
              "      <td></td>\n",
              "      <td>1666170534236</td>\n",
              "      <td>1.0</td>\n",
              "      <td>0</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>19Q4_primary-screen-mfi_NCIH661_LUNG</td>\n",
              "      <td></td>\n",
              "      <td></td>\n",
              "      <td>discover-test-datalake-v1</td>\n",
              "      <td>[NCI-H661]</td>\n",
              "      <td>[None]</td>\n",
              "      <td>[Carcinoma,  Large Cell]</td>\n",
              "      <td>[]</td>\n",
              "      <td>[]</td>\n",
              "      <td>[Homo Sapiens]</td>\n",
              "      <td>[]</td>\n",
              "      <td>[lung]</td>\n",
              "      <td></td>\n",
              "      <td>multiple_oa_2__19q4_primary_screen_mfi_ncih661...</td>\n",
              "      <td>0</td>\n",
              "      <td>Drug Screens</td>\n",
              "      <td>DEPMAP</td>\n",
              "      <td>Median fluorescence intensities for NCIH661_LU...</td>\n",
              "      <td>[]</td>\n",
              "      <td>https://discover-test-datalake-v1.s3-ap-southe...</td>\n",
              "      <td>gct</td>\n",
              "      <td>true</td>\n",
              "      <td>multiple_oa_2/data/220829-1818/19Q4_primary-sc...</td>\n",
              "      <td></td>\n",
              "      <td>multiple_oa_2/data</td>\n",
              "      <td></td>\n",
              "      <td></td>\n",
              "      <td>ap-southeast-1</td>\n",
              "      <td>multiple_oa_2</td>\n",
              "      <td>polly:data://multiple_oa_2/data/220829-1818/19...</td>\n",
              "      <td></td>\n",
              "      <td>1666170609899</td>\n",
              "      <td>1.0</td>\n",
              "      <td>0</td>\n",
              "      <td>2019</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>2019_Q3_Achilles_gene_effect_unscaled_GSS</td>\n",
              "      <td></td>\n",
              "      <td></td>\n",
              "      <td>discover-test-datalake-v1</td>\n",
              "      <td>[GSS]</td>\n",
              "      <td>[None]</td>\n",
              "      <td>[Stomach Neoplasms]</td>\n",
              "      <td>[]</td>\n",
              "      <td>[]</td>\n",
              "      <td>[Homo Sapiens]</td>\n",
              "      <td>[]</td>\n",
              "      <td>[stomach]</td>\n",
              "      <td></td>\n",
              "      <td>multiple_oa_2__2019_q3_achilles_gene_effect_un...</td>\n",
              "      <td>0</td>\n",
              "      <td>Gene dependency</td>\n",
              "      <td>DEPMAP</td>\n",
              "      <td>Post CERES file for GSS cell line, CERES data ...</td>\n",
              "      <td>[]</td>\n",
              "      <td>https://discover-test-datalake-v1.s3-ap-southe...</td>\n",
              "      <td>gct</td>\n",
              "      <td>true</td>\n",
              "      <td>multiple_oa_2/data/220829-1818/2019_Q3_Achille...</td>\n",
              "      <td></td>\n",
              "      <td>multiple_oa_2/data</td>\n",
              "      <td>CRISPR Screening</td>\n",
              "      <td>https://www.biorxiv.org/content/10.1101/720243v1</td>\n",
              "      <td>ap-southeast-1</td>\n",
              "      <td>multiple_oa_2</td>\n",
              "      <td>polly:data://multiple_oa_2/data/220829-1818/20...</td>\n",
              "      <td></td>\n",
              "      <td>1666170312873</td>\n",
              "      <td>1.0</td>\n",
              "      <td>0</td>\n",
              "      <td>2019</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>2019_Q3_Achilles_gene_effect_unscaled_HCC1937</td>\n",
              "      <td></td>\n",
              "      <td></td>\n",
              "      <td>discover-test-datalake-v1</td>\n",
              "      <td>[HCC1937]</td>\n",
              "      <td>[None]</td>\n",
              "      <td>[Breast Neoplasms]</td>\n",
              "      <td>[]</td>\n",
              "      <td>[]</td>\n",
              "      <td>[Homo Sapiens]</td>\n",
              "      <td>[]</td>\n",
              "      <td>[breast]</td>\n",
              "      <td></td>\n",
              "      <td>multiple_oa_2__2019_q3_achilles_gene_effect_un...</td>\n",
              "      <td>0</td>\n",
              "      <td>Gene dependency</td>\n",
              "      <td>DEPMAP</td>\n",
              "      <td>Post CERES file for HCC1937 cell line, CERES d...</td>\n",
              "      <td>[]</td>\n",
              "      <td>https://discover-test-datalake-v1.s3-ap-southe...</td>\n",
              "      <td>gct</td>\n",
              "      <td>true</td>\n",
              "      <td>multiple_oa_2/data/220714-1126/2019_Q3_Achille...</td>\n",
              "      <td></td>\n",
              "      <td>multiple_oa_2/data</td>\n",
              "      <td>CRISPR Screening</td>\n",
              "      <td>https://www.biorxiv.org/content/10.1101/720243v1</td>\n",
              "      <td>ap-southeast-1</td>\n",
              "      <td>multiple_oa_2</td>\n",
              "      <td>polly:data://multiple_oa_2/data/220714-1126/20...</td>\n",
              "      <td></td>\n",
              "      <td>1666170428323</td>\n",
              "      <td>1.0</td>\n",
              "      <td>0</td>\n",
              "      <td>2019</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>2020_Q2_Achilles_gene_effect_SUM159PT</td>\n",
              "      <td></td>\n",
              "      <td></td>\n",
              "      <td>discover-test-datalake-v1</td>\n",
              "      <td>[SUM159PT]</td>\n",
              "      <td>[None]</td>\n",
              "      <td>[Breast Neoplasms]</td>\n",
              "      <td>[]</td>\n",
              "      <td>[]</td>\n",
              "      <td>[Homo Sapiens]</td>\n",
              "      <td>[]</td>\n",
              "      <td>[breast]</td>\n",
              "      <td></td>\n",
              "      <td>multiple_oa_2__2020_q2_achilles_gene_effect_su...</td>\n",
              "      <td>0</td>\n",
              "      <td>Gene dependency</td>\n",
              "      <td>DEPMAP</td>\n",
              "      <td>Post CERES file for SUM159PT cell line, CERES ...</td>\n",
              "      <td>[]</td>\n",
              "      <td>https://discover-test-datalake-v1.s3-ap-southe...</td>\n",
              "      <td>gct</td>\n",
              "      <td>true</td>\n",
              "      <td>multiple_oa_2/data/220714-1126/2020_Q2_Achille...</td>\n",
              "      <td></td>\n",
              "      <td>multiple_oa_2/data</td>\n",
              "      <td>CRISPR Screening</td>\n",
              "      <td>https://www.biorxiv.org/content/10.1101/720243v1</td>\n",
              "      <td>ap-southeast-1</td>\n",
              "      <td>multiple_oa_2</td>\n",
              "      <td>polly:data://multiple_oa_2/data/220714-1126/20...</td>\n",
              "      <td></td>\n",
              "      <td>1666170489755</td>\n",
              "      <td>1.0</td>\n",
              "      <td>0</td>\n",
              "      <td>2020</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                                      dataset_id abstract author  \\\n",
              "0                                    01CO001_cnv                   \n",
              "1           19Q4_primary-screen-mfi_NCIH661_LUNG                   \n",
              "2      2019_Q3_Achilles_gene_effect_unscaled_GSS                   \n",
              "3  2019_Q3_Achilles_gene_effect_unscaled_HCC1937                   \n",
              "4          2020_Q2_Achilles_gene_effect_SUM159PT                   \n",
              "\n",
              "                      bucket curated_cell_line curated_cell_type  \\\n",
              "0  discover-test-datalake-v1            [None]            [None]   \n",
              "1  discover-test-datalake-v1        [NCI-H661]            [None]   \n",
              "2  discover-test-datalake-v1             [GSS]            [None]   \n",
              "3  discover-test-datalake-v1         [HCC1937]            [None]   \n",
              "4  discover-test-datalake-v1        [SUM159PT]            [None]   \n",
              "\n",
              "            curated_disease curated_drug curated_gene curated_organism  \\\n",
              "0       [Colonic Neoplasms]       [None]           []   [Homo Sapiens]   \n",
              "1  [Carcinoma,  Large Cell]           []           []   [Homo Sapiens]   \n",
              "2       [Stomach Neoplasms]           []           []   [Homo Sapiens]   \n",
              "3        [Breast Neoplasms]           []           []   [Homo Sapiens]   \n",
              "4        [Breast Neoplasms]           []           []   [Homo Sapiens]   \n",
              "\n",
              "  curated_strain      curated_tissue data_matrix_available  \\\n",
              "0             []  [colon sigmoideum]                         \n",
              "1             []              [lung]                         \n",
              "2             []           [stomach]                         \n",
              "3             []            [breast]                         \n",
              "4             []            [breast]                         \n",
              "\n",
              "                                     data_table_name  data_table_version  \\\n",
              "0                         multiple_oa_2__01co001_cnv                   0   \n",
              "1  multiple_oa_2__19q4_primary_screen_mfi_ncih661...                   0   \n",
              "2  multiple_oa_2__2019_q3_achilles_gene_effect_un...                   0   \n",
              "3  multiple_oa_2__2019_q3_achilles_gene_effect_un...                   0   \n",
              "4  multiple_oa_2__2020_q2_achilles_gene_effect_su...                   0   \n",
              "\n",
              "                   data_type dataset_source  \\\n",
              "0  ['copy number variation']          CPTAC   \n",
              "1               Drug Screens         DEPMAP   \n",
              "2            Gene dependency         DEPMAP   \n",
              "3            Gene dependency         DEPMAP   \n",
              "4            Gene dependency         DEPMAP   \n",
              "\n",
              "                                         description drug_smiles  \\\n",
              "0  Copy number variation values for 01CO001 in co...          []   \n",
              "1  Median fluorescence intensities for NCIH661_LU...          []   \n",
              "2  Post CERES file for GSS cell line, CERES data ...          []   \n",
              "3  Post CERES file for HCC1937 cell line, CERES d...          []   \n",
              "4  Post CERES file for SUM159PT cell line, CERES ...          []   \n",
              "\n",
              "                                       file_location file_type is_current  \\\n",
              "0  https://discover-test-datalake-v1.s3-ap-southe...       gct       true   \n",
              "1  https://discover-test-datalake-v1.s3-ap-southe...       gct       true   \n",
              "2  https://discover-test-datalake-v1.s3-ap-southe...       gct       true   \n",
              "3  https://discover-test-datalake-v1.s3-ap-southe...       gct       true   \n",
              "4  https://discover-test-datalake-v1.s3-ap-southe...       gct       true   \n",
              "\n",
              "                                                 key overall_design  \\\n",
              "0     multiple_oa_2/data/220829-1818/01CO001_cnv.gct                  \n",
              "1  multiple_oa_2/data/220829-1818/19Q4_primary-sc...                  \n",
              "2  multiple_oa_2/data/220829-1818/2019_Q3_Achille...                  \n",
              "3  multiple_oa_2/data/220714-1126/2019_Q3_Achille...                  \n",
              "4  multiple_oa_2/data/220714-1126/2020_Q2_Achille...                  \n",
              "\n",
              "              package          platform  \\\n",
              "0  multiple_oa_2/data                     \n",
              "1  multiple_oa_2/data                     \n",
              "2  multiple_oa_2/data  CRISPR Screening   \n",
              "3  multiple_oa_2/data  CRISPR Screening   \n",
              "4  multiple_oa_2/data  CRISPR Screening   \n",
              "\n",
              "                                        publication          region  \\\n",
              "0        https://doi.org/10.1016/j.cell.2019.03.030  ap-southeast-1   \n",
              "1                                                    ap-southeast-1   \n",
              "2  https://www.biorxiv.org/content/10.1101/720243v1  ap-southeast-1   \n",
              "3  https://www.biorxiv.org/content/10.1101/720243v1  ap-southeast-1   \n",
              "4  https://www.biorxiv.org/content/10.1101/720243v1  ap-southeast-1   \n",
              "\n",
              "        src_repo                                            src_uri summary  \\\n",
              "0  multiple_oa_2  polly:data://multiple_oa_2/data/220829-1818/01...           \n",
              "1  multiple_oa_2  polly:data://multiple_oa_2/data/220829-1818/19...           \n",
              "2  multiple_oa_2  polly:data://multiple_oa_2/data/220829-1818/20...           \n",
              "3  multiple_oa_2  polly:data://multiple_oa_2/data/220714-1126/20...           \n",
              "4  multiple_oa_2  polly:data://multiple_oa_2/data/220714-1126/20...           \n",
              "\n",
              "      timestamp_  total_num_samples  version  year  \n",
              "0  1666170534236                1.0        0        \n",
              "1  1666170609899                1.0        0  2019  \n",
              "2  1666170312873                1.0        0  2019  \n",
              "3  1666170428323                1.0        0  2019  \n",
              "4  1666170489755                1.0        0  2020  "
            ]
          },
          "execution_count": 12,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# ----------- QUERY EXAMPLE 1 -------------- # Query all datasets from GEO\n",
        "\n",
        "query = \"\"\"SELECT * FROM multiple_oa_2.geo.datasets\"\"\"\n",
        "results=omixatlas.query_metadata(query)\n",
        "results.head()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "# ----------- QUERY EXAMPLE 2 -------------- # Query all datasets from CBioPortal\n",
        "\n",
        "query = \"\"\"SELECT * FROM multiple_oa_2.cbioportal.datasets\"\"\"\n",
        "results=omixatlas.query_metadata(query)\n",
        "results.head()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "# ----------- QUERY EXAMPLE 3 -------------- # Query mutation datasets from CBioPortal\n",
        "\n",
        "query = \"\"\"SELECT * FROM multiple_oa_2.cbioportal.mutation.datasets\"\"\"\n",
        "results=omixatlas.query_metadata(query)\n",
        "results.head()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "# ----------- QUERY EXAMPLE 4 -------------- # Query transcriptomics datasets from GEO\n",
        "\n",
        "query = \"\"\"SELECT * FROM multiple_oa_2.geo.transcriptomics.datasets\"\"\"\n",
        "results=omixatlas.query_metadata(query)\n",
        "results.head()"
      ]
    }
  ],
  "metadata": {
    "kernelspec": {
      "display_name": "Python 3 (ipykernel)",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.10.2"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 2
}