{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "6a58cca0",
   "metadata": {},
   "source": [
    "# PollyGraph : Answering biomedical questions with large biomedical knowledge graphs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ed7ccfd7",
   "metadata": {},
   "source": [
    "<img src=\"images/logo-logo.jpg\" alt=\"Drawing\" style=\"width: 500px;\"/>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "83e712b4",
   "metadata": {},
   "source": [
    "### What is PollyGraph ?\n",
    "\n",
    "PollyGraph a.k.a ontology-driven knowledge graph is a semantic model for biomedical molecular data. \n",
    "\n",
    "In most organisations, big and small, information is spread out across different systems in variety of formats and contexts. As a result, users spend much time trying to manually find relevant data for their use. \n",
    "\n",
    "Polly, which is home to over 1.6 million biomedical molecular data, is standardised for both data and metadata. There are two major advantages to this effort. First, data standardisation using uniform file formats such as GCT, h5ad, vcf etc. enable large scale consumption of data for analysis. Second, metadata harmonisation enables users to perform data findability at scale without ever worrying to provide exact keywords for search.\n",
    "\n",
    "To make search even more powerful for scientists, we introduce PollyGraph to perform semantic searches using several biological concepts.\n",
    "\n",
    "### What is Semantic Search ? \n",
    "\n",
    "Semantic search is an advanced technology for optimizing the accuracy of our search results by enabling models to understand the meaning of concepts and the relationships between them. \n",
    "\n",
    "### Why do we need semantics ? \n",
    "\n",
    "As the data on Polly is rapidly growing in number, more and more knowledge populates the platform. Therefore, it becomes increasingly challenging for machines to process and retrieve \"relevant\" information on our behalf. Though it is easy for humans to decide whether two or more entities are associated, machines struggle and often fail to do it. \n",
    "\n",
    "### The perks of semantic-search\n",
    "\n",
    "- It's context-aware : It identities entities that are relevant for the task and retreieves relevant datasets.\n",
    "- It's extensively interlinked : Provides multiple references for an entitiy across multiple domain specific concepts and knowledge grows significantly by adding new associations between entities\n",
    "- It's highly personalised : Capable of returning results that are more closely aligned with user's interests."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "74812d9c",
   "metadata": {},
   "source": [
    "<img src=\"images/onto-text.png\" alt=\"Drawing\" style=\"width: 500px;\"/>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "12b237c9",
   "metadata": {},
   "source": [
    "## Ontology-driven Biomedical Knowledge Graphs\n",
    "\n",
    "One of the main mechanisms behind this ability of semantic search is to utilise publicly available biomedical ontologies to provide more meaningful results using knowledge graph. Linking biomedical ontologies (e.g. MeSH, BTO, CVCL, HGNC, GO etc) by mining associations between them.\n",
    "\n",
    "By leveraging ontologies, semantic search is able to provide a suitable response even if the results don’t contain the exact wording of the query."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c3765349",
   "metadata": {},
   "source": [
    "<img src=\"images/primary_model.jpg\" alt=\"model\" style=\"width: 800px;\"/>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fd1a91b0",
   "metadata": {},
   "source": [
    "### 1. Connecting to PollyGraph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "737a2013",
   "metadata": {},
   "outputs": [],
   "source": [
    "from neo4j import GraphDatabase\n",
    "import pandas as pd\n",
    "import tqdm\n",
    "import os\n",
    "import time\n",
    "from polly.auth import Polly\n",
    "from polly.omixatlas import OmixAtlas\n",
    "\n",
    "AUTH_TOKEN=\"eyJjdHkiOiJKV1QiLCJlbmMiOiJBMjU2R0NNIiwiYWxnIjoiUlNBLU9BRVAifQ.QzHksBx2ZUfXnqD3o7EVj56A7WHzZ8fOkQ72_eCmef1igtab8OevFsU7Hs-TmfOPc2BkOn2WEEuj0l6Gq2_LRWpsJWkgQyw1yHovJyWTo9ntlvCD3qFsmrizIpaACsqoLG2wuf77RlIFKBA4Tr9G5EfntlWDXV9nmLhU1dnl7iBafiHfhFNOJhCKgEroMRhczr8Hqqw-iMqbn9QDcR1rrTp8cSYjUVq2PAVVnr90mwirV4Cv3Rs8WIeL8yNOhnzurLVmBAXs6f5Pe0DyNeg_DpureaM2UUsGZLPHbPsprpdqHf78Ffdl7mab6-cNgKTa1NwLVW36LWHZFQJAfsZ8Uw.msX5rQydqbU0I4sh.IjkX_ypX4hBitWFQZgty0OsebIlbMeolwGPqV2EyOYzHVLMJi6Wu_MO7yAptw3K9OakZ7J3OcG4QLJ_SlqU__tT0dx_AH13otXqqSBVDdPGEq3FAdyKuMwvnX5yQxo8gIhjadNlu67eXansia61GqQOICXi8L9gGb49K9JttEC8Bsb_btHZuNEk-rIFJn838SdB-bo0aQzP2K1OgbuUrX92ttvsZvo8j5h8g-gakRRq6DEpV86jaoZ0PrOzSEnjOv0iP6qWpUyPvr_JJgIY63K243WYLJfgl9K_TtQvkr-k2OcKoCpiIxnN_GLIiDnk9XnqtiIycwpTN8SsB7bn5xjaKKjumrh-m6vHlz6vbFH3qycRY2N0JGPUA9Iz1UZ0niYNx8OIRW635Lj7KYN5l56Ob8PIwvMVEEKzavILYPasdcvIJGWXQKBjyGZIs-ML6OOohWMSm4PvnoPAtRcAdm2axpp4TNfLpErlBPiBcEEml50XxI6Pwx2hpCD46OeGYZbNOjeyLA7asrJmVpqIo_j5Ga2wqd7nxPeo_d4vu4r_eKAlh28GZ_EewBpQ9CbJW54Belpr6BxCYczp_uht3pHLSI45Tu5OI6r6KXZ3XOvN7UxthOCpdZn7t7VE0YUIrFAgZFVPjMu9FqEdbjfZ_Ueceh26GviT-Mv87CdhAuLBuScoyzCAV6UYRfoV62il1r3LyDr5WriZiycnM-BzGAQsA22qOOJYOWhsMDYZQ_AvKvbbwkYNBmMk6x77ul5sduieKvg-c04QfZ5cGOns-E0NBO-P1AHdwbZkjDQzcyns7dot7f_qFscm5XlOJGB3yXWuv2_DrZz1BQYhjn2ZARG3YQkdWChsap93f6GWjoV_qMDIZ8YwjsiaToNrBH9wSA6JdxGiWS5DZPPQSVK_vMFBHQnxGHaKfb6t2hIDk2Qr4ruc15ggLEuU9LL7iwbLR69iUevk-CUtT2iVBSH608iDkgH1y5vgecLyJMYK8QnFpHR_xnaqi6qF8nrTK4O060SZQbZkop7Lx-Wxv74P2_hXMbsEcH_gW0PaWd8Id_MYKrK4q-oIaJ_ghyERbiOJVqgLtKn5U7ltsmbm1RnSZROoNeoaLDBTdlerG1o8R7D90rJL0hh30kQY2-thrnVUceZkSWAYTZ48wyU8W08OSwUdZ13fRYO96MyEiGtuBoNmVIZ0pewtH04YCGHNXPQYwJ_jqEXTYXCh6jXe11JFQDprK-2r22iMubEdJ6HH_r1FzcX3XbBG4dNcfp1q9pM_hvp-48MVXfYosfp2nNIKfZERMGrulNDUIMJJJAjb16qh-5vfNb7lIQd0G210.7-0XNxC4nnp-FL1SmIHvGA\" # Obtain authentication tokens\n",
    "Polly.auth(AUTH_TOKEN)\n",
    "\n",
    "#Defining omixatlas object\n",
    "omixatlas = OmixAtlas()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "6d1092e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "class PollyGraph:\n",
    "    \n",
    "    def __init__(self, uri, user, pwd):\n",
    "        self.__uri = uri\n",
    "        self.__user = user\n",
    "        self.__pwd = pwd\n",
    "        self.__driver = None\n",
    "        try:\n",
    "            self.__driver = GraphDatabase.driver(self.__uri, auth=(self.__user, self.__pwd))\n",
    "        except Exception as e:\n",
    "            print(\"Failed to create the driver:\", e)\n",
    "        \n",
    "    def close(self):\n",
    "        if self.__driver is not None:\n",
    "            self.__driver.close()\n",
    "        \n",
    "    def query(self, query, parameters=None, db=None):\n",
    "        assert self.__driver is not None, \"Driver not initialized!\"\n",
    "        session = None\n",
    "        response = None\n",
    "        try: \n",
    "            session = self.__driver.session(database=db) if db is not None else self.__driver.session() \n",
    "            response = list(session.run(query, parameters))\n",
    "        except Exception as e:\n",
    "            print(\"Query failed:\", e)\n",
    "        finally: \n",
    "            if session is not None:\n",
    "                session.close()\n",
    "        return response\n",
    "    \n",
    "def build_rel_query(field_name, col_name, meta_field, rel):\n",
    "    \n",
    "    if rel == 'mentioned_in':\n",
    "        q = \"\"\"\n",
    "            UNWIND $rows as row\n",
    "            MATCH (n:ns0__\"\"\"+field_name +\"\"\"{ns0__\"\"\"+meta_field+\"\"\": [row.\"\"\"+col_name+\"\"\"]})\n",
    "            MATCH (d:dataset {src_dataset_id: row.src_dataset_id})\n",
    "            MERGE (n)-[:\"\"\"+rel+\"\"\"]->(d)\n",
    "            RETURN count(*) as total\n",
    "            \"\"\"\n",
    "    else:\n",
    "        q = \"\"\"\n",
    "            UNWIND $rows as row\n",
    "            MATCH (n:ns0__\"\"\"+field_name +\"\"\"{ns0__\"\"\"+meta_field+\"\"\": row.\"\"\"+col_name+\"\"\"})\n",
    "            MATCH (d:dataset {src_dataset_id: row.src_dataset_id})\n",
    "            MERGE (n)-[:\"\"\"+rel+\"\"\"]->(d)\n",
    "            RETURN count(*) as total\n",
    "            \"\"\"\n",
    "        \n",
    "    return(q)\n",
    "    \n",
    "def add_dataset(rows, batch_size=500):\n",
    "   # Adds category nodes to the Neo4j graph.\n",
    "    query = '''\n",
    "            UNWIND $rows AS row\n",
    "            MERGE (c:dataset {\n",
    "                dataset_id: row.dataset_id,\n",
    "                src_dataset_id: row.src_dataset_id, \n",
    "                src_overall_design: row.src_overall_design,\n",
    "                src_description: row.src_description,\n",
    "                src_summary: row.src_summary,\n",
    "                data_type: row.data_type,\n",
    "                curated_cell_line: row.curated_cell_line,\n",
    "                curated_cell_type: row.curated_cell_type,\n",
    "                curated_disease: row.curated_disease,\n",
    "                curated_drug: row.curated_drug,\n",
    "                curated_gene: row.curated_gene,\n",
    "                curated_tissue: row.curated_tissue,\n",
    "                condition_column: row.condition_column,\n",
    "                condition_control: row.condition_control,\n",
    "                condition_perturbation: row.condition_perturbation\n",
    "                })\n",
    "            \n",
    "            RETURN count(*) as total\n",
    "            '''\n",
    "    return insert_data(query, rows)\n",
    "\n",
    "def insert_data(query, rows, batch_size = 500):\n",
    "    # Function to handle the updating the Neo4j database in batch mode.\n",
    "\n",
    "    total = 0\n",
    "    batch = 0\n",
    "    start = time.time()\n",
    "    result = None\n",
    "\n",
    "    while batch * batch_size < len(rows):\n",
    "        res = pollygraph.query(query, \n",
    "                         parameters = {'rows': rows[batch*batch_size:(batch+1)*batch_size].to_dict('records')})\n",
    "        total += res[0]['total']\n",
    "        batch += 1\n",
    "        result = {\"total\":total, \n",
    "                  \"batches\":batch, \n",
    "                  \"time\":time.time()-start}\n",
    "        #print(result)\n",
    "\n",
    "    return result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "235efe61",
   "metadata": {},
   "outputs": [],
   "source": [
    "pollygraph = PollyGraph(uri=\"bolt://localhost:7687\", user=\"neo4j\", pwd=\"password\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1af092cf",
   "metadata": {},
   "source": [
    "### 2. Get dataset metadata from Polly"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "fb7522ca",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Query execution succeeded (time taken: 2.81 seconds, data scanned: 2.382 MB)\n",
      "Fetched 5120 rows\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dataset_id</th>\n",
       "      <th>src_dataset_id</th>\n",
       "      <th>condition_column</th>\n",
       "      <th>condition_control</th>\n",
       "      <th>condition_perturbation</th>\n",
       "      <th>curated_cell_line</th>\n",
       "      <th>curated_cell_type</th>\n",
       "      <th>curated_disease</th>\n",
       "      <th>curated_drug</th>\n",
       "      <th>curated_gene</th>\n",
       "      <th>curated_tissue</th>\n",
       "      <th>data_type</th>\n",
       "      <th>src_description</th>\n",
       "      <th>src_overall_design</th>\n",
       "      <th>src_summary</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GSE100054_GPL23126-2022-06-21-07-47-56</td>\n",
       "      <td>GSE100054_GPL23126</td>\n",
       "      <td>kw_curated_disease</td>\n",
       "      <td>[Normal]</td>\n",
       "      <td>[Parkinson Disease]</td>\n",
       "      <td>[None]</td>\n",
       "      <td>[neuron, peripheral blood mononuclear cell]</td>\n",
       "      <td>[Parkinson Disease]</td>\n",
       "      <td>[None]</td>\n",
       "      <td>[AMBRA1, HDAC6, BECN1, ULK1, ATG5, ATG4B, SNCA...</td>\n",
       "      <td>[brain, blood]</td>\n",
       "      <td>Transcriptomics</td>\n",
       "      <td>Expression profiling of peripheral blood monon...</td>\n",
       "      <td>PBMC in 9 normal controls and 10 patients with...</td>\n",
       "      <td>Autophagy is a highly conserved degradation pa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GSE10006_GPL570-2022-06-21-07-47-57</td>\n",
       "      <td>GSE10006_GPL570</td>\n",
       "      <td>kw_curated_disease</td>\n",
       "      <td>[Normal]</td>\n",
       "      <td>[Pulmonary Disease, Chronic Obstructive]</td>\n",
       "      <td>[None]</td>\n",
       "      <td>[secretory cell, respiratory epithelial cell]</td>\n",
       "      <td>[Bacterial Infections, Pulmonary Disease, Chro...</td>\n",
       "      <td>[None]</td>\n",
       "      <td>[ITLN1]</td>\n",
       "      <td>[None]</td>\n",
       "      <td>Transcriptomics</td>\n",
       "      <td>Decreased Expression of Intelectin 1 in The Hu...</td>\n",
       "      <td>Comparison of gene expression in airway epithe...</td>\n",
       "      <td>Lectins are proteins present on cell surfaces ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GSE100095_GPL17586-2022-07-08-05-43-08</td>\n",
       "      <td>GSE100095_GPL17586</td>\n",
       "      <td>kw_curated_drug</td>\n",
       "      <td>[none]</td>\n",
       "      <td>[3',5'-cyclic AMP]</td>\n",
       "      <td>[BeWo]</td>\n",
       "      <td>[None]</td>\n",
       "      <td>[Choriocarcinoma]</td>\n",
       "      <td>[3',5'-cyclic AMP, progesterone]</td>\n",
       "      <td>[None]</td>\n",
       "      <td>[None]</td>\n",
       "      <td>Transcriptomics</td>\n",
       "      <td>Expression data from BeWo cells treated withou...</td>\n",
       "      <td>Human placenta choriocarcinoma cell line BeWo ...</td>\n",
       "      <td>Cyclic AMP activates two downstream factors, p...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GSE100194_GPL7202-2022-07-08-05-43-09</td>\n",
       "      <td>GSE100194_GPL7202</td>\n",
       "      <td>kw_curated_drug</td>\n",
       "      <td>[none]</td>\n",
       "      <td>[ketotifen]</td>\n",
       "      <td>[None]</td>\n",
       "      <td>[mast cell]</td>\n",
       "      <td>[Dengue, Flavivirus Infections]</td>\n",
       "      <td>[ketotifen]</td>\n",
       "      <td>[None]</td>\n",
       "      <td>[liver, spleen]</td>\n",
       "      <td>Transcriptomics</td>\n",
       "      <td>Transcriptional Profiling Confirms the Therape...</td>\n",
       "      <td>Gene expression in liver and spleen were measu...</td>\n",
       "      <td>In this study, we show that the host response ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GSE100195_GPL7202-2022-07-08-05-43-10</td>\n",
       "      <td>GSE100195_GPL7202</td>\n",
       "      <td>kw_curated_drug</td>\n",
       "      <td>[none]</td>\n",
       "      <td>[ketotifen]</td>\n",
       "      <td>[None]</td>\n",
       "      <td>[mast cell]</td>\n",
       "      <td>[Seizures, Febrile, Hemorrhagic Disorders, Vas...</td>\n",
       "      <td>[ketotifen]</td>\n",
       "      <td>[None]</td>\n",
       "      <td>[spleen, liver]</td>\n",
       "      <td>Transcriptomics</td>\n",
       "      <td>Transcriptional Profiling Confirms the Therape...</td>\n",
       "      <td>Gene expression in liver and spleen were measu...</td>\n",
       "      <td>In this study, we show that the host response ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                               dataset_id      src_dataset_id  \\\n",
       "0  GSE100054_GPL23126-2022-06-21-07-47-56  GSE100054_GPL23126   \n",
       "1     GSE10006_GPL570-2022-06-21-07-47-57     GSE10006_GPL570   \n",
       "2  GSE100095_GPL17586-2022-07-08-05-43-08  GSE100095_GPL17586   \n",
       "3   GSE100194_GPL7202-2022-07-08-05-43-09   GSE100194_GPL7202   \n",
       "4   GSE100195_GPL7202-2022-07-08-05-43-10   GSE100195_GPL7202   \n",
       "\n",
       "     condition_column condition_control  \\\n",
       "0  kw_curated_disease          [Normal]   \n",
       "1  kw_curated_disease          [Normal]   \n",
       "2     kw_curated_drug            [none]   \n",
       "3     kw_curated_drug            [none]   \n",
       "4     kw_curated_drug            [none]   \n",
       "\n",
       "                     condition_perturbation curated_cell_line  \\\n",
       "0                       [Parkinson Disease]            [None]   \n",
       "1  [Pulmonary Disease, Chronic Obstructive]            [None]   \n",
       "2                        [3',5'-cyclic AMP]            [BeWo]   \n",
       "3                               [ketotifen]            [None]   \n",
       "4                               [ketotifen]            [None]   \n",
       "\n",
       "                               curated_cell_type  \\\n",
       "0    [neuron, peripheral blood mononuclear cell]   \n",
       "1  [secretory cell, respiratory epithelial cell]   \n",
       "2                                         [None]   \n",
       "3                                    [mast cell]   \n",
       "4                                    [mast cell]   \n",
       "\n",
       "                                     curated_disease  \\\n",
       "0                                [Parkinson Disease]   \n",
       "1  [Bacterial Infections, Pulmonary Disease, Chro...   \n",
       "2                                  [Choriocarcinoma]   \n",
       "3                    [Dengue, Flavivirus Infections]   \n",
       "4  [Seizures, Febrile, Hemorrhagic Disorders, Vas...   \n",
       "\n",
       "                       curated_drug  \\\n",
       "0                            [None]   \n",
       "1                            [None]   \n",
       "2  [3',5'-cyclic AMP, progesterone]   \n",
       "3                       [ketotifen]   \n",
       "4                       [ketotifen]   \n",
       "\n",
       "                                        curated_gene   curated_tissue  \\\n",
       "0  [AMBRA1, HDAC6, BECN1, ULK1, ATG5, ATG4B, SNCA...   [brain, blood]   \n",
       "1                                            [ITLN1]           [None]   \n",
       "2                                             [None]           [None]   \n",
       "3                                             [None]  [liver, spleen]   \n",
       "4                                             [None]  [spleen, liver]   \n",
       "\n",
       "         data_type                                    src_description  \\\n",
       "0  Transcriptomics  Expression profiling of peripheral blood monon...   \n",
       "1  Transcriptomics  Decreased Expression of Intelectin 1 in The Hu...   \n",
       "2  Transcriptomics  Expression data from BeWo cells treated withou...   \n",
       "3  Transcriptomics  Transcriptional Profiling Confirms the Therape...   \n",
       "4  Transcriptomics  Transcriptional Profiling Confirms the Therape...   \n",
       "\n",
       "                                  src_overall_design  \\\n",
       "0  PBMC in 9 normal controls and 10 patients with...   \n",
       "1  Comparison of gene expression in airway epithe...   \n",
       "2  Human placenta choriocarcinoma cell line BeWo ...   \n",
       "3  Gene expression in liver and spleen were measu...   \n",
       "4  Gene expression in liver and spleen were measu...   \n",
       "\n",
       "                                         src_summary  \n",
       "0  Autophagy is a highly conserved degradation pa...  \n",
       "1  Lectins are proteins present on cell surfaces ...  \n",
       "2  Cyclic AMP activates two downstream factors, p...  \n",
       "3  In this study, we show that the host response ...  \n",
       "4  In this study, we show that the host response ...  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "q = \"\"\"SELECT \n",
    "              dataset_id,\n",
    "              src_dataset_id, \n",
    "              src_overall_design,\n",
    "              src_description,\n",
    "              src_summary,\n",
    "              data_type,\n",
    "              curated_cell_line,\n",
    "              curated_cell_type,\n",
    "              curated_disease,\n",
    "              curated_drug,\n",
    "              curated_gene,\n",
    "              curated_tissue,\n",
    "              condition_column,\n",
    "              condition_control,\n",
    "              condition_perturbation\n",
    "       FROM gdx_atlas.datasets\"\"\"\n",
    "\n",
    "df = omixatlas.query_metadata(q, query_api_version='v2')\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fd745836",
   "metadata": {},
   "source": [
    "### 3. Integrate Dataset Metadata on Polly with KG"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "8827eaba",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'total': 5120, 'batches': 11, 'time': 5.165494680404663}"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "add_dataset(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "726d5968",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Adding relationships for each dataset to biological entities\n",
    "curated_columns = ['curated_cell_line', 'curated_cell_type',\n",
    "                  'curated_drug', 'curated_tissue', 'curated_gene']\n",
    "\n",
    "for index, row in tqdm.tqdm(df.iterrows()):\n",
    "    index += 1\n",
    "    row = df[index-1:index]\n",
    "    pert_col = 'condition_perturbation'\n",
    "    \n",
    "    if row['condition_column'][index-1] == 'kw_curated_disease' or row['condition_column'][index-1] == 'kw_curated_drug':\n",
    "        pert_field = row['condition_column'][index-1].split('_')[-1]+' '\n",
    "        query = build_rel_query(pert_field, pert_col, \"name\", \"is_perturbed_in\")\n",
    "        insert_data(query, row)\n",
    "    \n",
    "    for col in curated_columns:\n",
    "        field_name = col.replace('curated_', '')\n",
    "        item = row[col][index-1]\n",
    "        row_df = row.explode(col)\n",
    "        if col == 'curated_gene' and \"\".join(item).lower() != \"none\":\n",
    "            query = build_rel_query(field_name, col, \"gene_symbol\", \"mentioned_in\")\n",
    "            insert_data(query, row_df)\n",
    "            \n",
    "        elif col != 'curated_gene' and \"\".join(item).lower() != \"none\":\n",
    "            query = build_rel_query(field_name, col, \"name\", \"mentioned_in\")\n",
    "            insert_data(query, row_df)\n",
    "            \n",
    "        else:\n",
    "            continue"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "706c44c5",
   "metadata": {},
   "source": [
    "## Context-driven search examples using PollyGraph"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "55bc8fe0",
   "metadata": {},
   "source": [
    "### Q1: List genes that constitute TGF-β receptor signaling pathway that are differentially expressed in Prostate Cancer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "cc7e2625",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>pathway</th>\n",
       "      <th>disease</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>[ALKBH5]</td>\n",
       "      <td>[Negative regulation of transforming growth fa...</td>\n",
       "      <td>[Prostatic Neoplasms]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>[TSPY26P]</td>\n",
       "      <td>[Negative regulation of transforming growth fa...</td>\n",
       "      <td>[Prostatic Neoplasms]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>[METTL1]</td>\n",
       "      <td>[Negative regulation of transforming growth fa...</td>\n",
       "      <td>[Prostatic Neoplasms]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>[STIL]</td>\n",
       "      <td>[Negative regulation of transforming growth fa...</td>\n",
       "      <td>[Prostatic Neoplasms]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>[ROR2]</td>\n",
       "      <td>[Negative regulation of transforming growth fa...</td>\n",
       "      <td>[Prostatic Neoplasms]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>821</th>\n",
       "      <td>[TRIM24]</td>\n",
       "      <td>[Transforming growth factor beta receptor sign...</td>\n",
       "      <td>[Prostatic Neoplasms]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>822</th>\n",
       "      <td>[DMBT1]</td>\n",
       "      <td>[Transforming growth factor beta receptor sign...</td>\n",
       "      <td>[Prostatic Neoplasms]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>823</th>\n",
       "      <td>[ARG1]</td>\n",
       "      <td>[Transforming growth factor beta receptor sign...</td>\n",
       "      <td>[Prostatic Neoplasms]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>824</th>\n",
       "      <td>[DUS3L]</td>\n",
       "      <td>[Transforming growth factor beta receptor sign...</td>\n",
       "      <td>[Prostatic Neoplasms]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>825</th>\n",
       "      <td>[DLX2]</td>\n",
       "      <td>[Transforming growth factor beta receptor sign...</td>\n",
       "      <td>[Prostatic Neoplasms]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>826 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          gene                                            pathway  \\\n",
       "0     [ALKBH5]  [Negative regulation of transforming growth fa...   \n",
       "1    [TSPY26P]  [Negative regulation of transforming growth fa...   \n",
       "2     [METTL1]  [Negative regulation of transforming growth fa...   \n",
       "3       [STIL]  [Negative regulation of transforming growth fa...   \n",
       "4       [ROR2]  [Negative regulation of transforming growth fa...   \n",
       "..         ...                                                ...   \n",
       "821   [TRIM24]  [Transforming growth factor beta receptor sign...   \n",
       "822    [DMBT1]  [Transforming growth factor beta receptor sign...   \n",
       "823     [ARG1]  [Transforming growth factor beta receptor sign...   \n",
       "824    [DUS3L]  [Transforming growth factor beta receptor sign...   \n",
       "825     [DLX2]  [Transforming growth factor beta receptor sign...   \n",
       "\n",
       "                   disease  \n",
       "0    [Prostatic Neoplasms]  \n",
       "1    [Prostatic Neoplasms]  \n",
       "2    [Prostatic Neoplasms]  \n",
       "3    [Prostatic Neoplasms]  \n",
       "4    [Prostatic Neoplasms]  \n",
       "..                     ...  \n",
       "821  [Prostatic Neoplasms]  \n",
       "822  [Prostatic Neoplasms]  \n",
       "823  [Prostatic Neoplasms]  \n",
       "824  [Prostatic Neoplasms]  \n",
       "825  [Prostatic Neoplasms]  \n",
       "\n",
       "[826 rows x 3 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query_string = '''\n",
    "MATCH (n:ns0__pathway)--(m:ns0__disease)--(p:ns0__gene)\n",
    "WHERE ANY (x in n.ns0__name WHERE toLower(x) CONTAINS(\"transforming growth factor beta receptor signaling\")) AND\n",
    "ANY (y in m.ns0__name+m.ns0__synonyms WHERE toLower(y) CONTAINS(\"prostate cancer\"))\n",
    "RETURN p.ns0__gene_symbol as gene, n.ns0__name as pathway, m.ns0__name as disease;\n",
    "'''\n",
    "top_cat_df = pd.DataFrame([dict(_) for _ in pollygraph.query(query_string)])\n",
    "top_cat_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cb19f67f",
   "metadata": {},
   "source": [
    "### Q2: Find datasets related to Brain Injury where BRD1 gene is regulated"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "de6a9f50",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>src_dataset_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GSE68207_GPL18694</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GSE92363_GPL14746</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GSE45997_GPL1355</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GSE44625_GPL6885</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GSE41345_GPL6885</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>GSE39759_GPL7202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>GSE58613_GPL571</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>GSE116980_GPL22396</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>GSE111452_GPL15084</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       src_dataset_id\n",
       "0   GSE68207_GPL18694\n",
       "1   GSE92363_GPL14746\n",
       "2    GSE45997_GPL1355\n",
       "3    GSE44625_GPL6885\n",
       "4    GSE41345_GPL6885\n",
       "5    GSE39759_GPL7202\n",
       "6     GSE58613_GPL571\n",
       "7  GSE116980_GPL22396\n",
       "8  GSE111452_GPL15084"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query_string = '''\n",
    "MATCH (n:ns0__disease) WHERE ANY (x in n.ns0__name + n.ns0__synonyms where toLower(x) CONTAINS(\"brain injury\")) \n",
    "WITH n.ns0__name as disease\n",
    "UNWIND disease as d\n",
    "MATCH (m:ns0__gene {ns0__gene_symbol : ['BRD1']})--(n:ns0__disease {ns0__name : [d]})--(p: dataset)\n",
    "RETURN p.src_dataset_id as src_dataset_id;\n",
    "'''\n",
    "top_cat_df = pd.DataFrame([dict(_) for _ in pollygraph.query(query_string)])\n",
    "top_cat_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ff649771",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "general",
   "language": "python",
   "name": "general"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
